---
title: "DESeq2 results"
author: "CCBR RNAseq pipeline"
output: html_document
params:
  folder: result
  sampleinfo: sampletable.txt
  data: RawCountFile_genes_filtered.txt
  contrasts: "GN GP"
  species: hg19
  projectId: "<projectId>"
  projectDesc: "<desc>"
  gtffile: "gencode.v19.annotation.gtf"
  dtype: "<count type>"
  karyobeds: "path/to/karyobedfiles"
---

```{r, include=FALSE, warning=FALSE, message=FALSE}
countMethod <- "subread"
if(grepl("RSEM",params$data)){
  countMethod <- "rsem"
}
projectId<-params$projectId
projectDesc<-params$projectDesc
dateandtime<-format(Sys.time(), "%a %b %d %Y - %X")
```
### **Project:**
####    *`r projectId`*
### **Description:** 
####    *`r projectDesc`*
###  **Count method:** 
####    *`r countMethod`* 
### **Report generated:** 
####    *`r dateandtime`* 

## **Histogram after normalization**

```{r setup, echo=FALSE, warning=FALSE,message=FALSE}

library(rgl)
library(knitr)
library('DESeq2')
library('statmod')
library('RColorBrewer')
library('gplots')
library('reshape')
library('ggplot2')
library('ggfortify')
library(amap)
library(DT)
library(plotly)
library('geneplotter')

knit_hooks$set(rgl = function(before, options, envir) {
  if (!before) {
    ## after a chunk has been evaluated
    if (rgl.cur() == 0) return()  # no active device
    name = paste(options$fig.path, options$label, sep = '')
    rgl.snapshot(paste(name, '.png', sep = ''), fmt = 'png')
    return(paste('\\includegraphics{', name, '}\n', sep = ''))
  }
})

knit_hooks$set(webgl = hook_webgl)
```

```{r, echo=FALSE, warning=FALSE,message=FALSE}

## grab args & normalization

DIR <- params$folder
FILE1 <- params$sampleinfo
FILE2 <- params$data
CONTRASTS <- params$contrasts
SPECIES <-params$species
KARYOBEDS <-params$karyobeds
# ANNOTATE <- args[4]

#
## setwd(DIR)
## Sys.setenv("DISPLAY"=":0.0")
sampleinfo=read.delim(FILE1)
sampleFiles=as.character(sampleinfo[,2])
x = read.delim(FILE2,row.names=1)
colnames(x)=as.character(sampleinfo[,4])
## read annotation file
## ann=read.delim(ANNOTATE)
#
ddsHTSeq<-DESeqDataSetFromMatrix(countData=x,colData=sampleinfo, design=~condition)
dds<-DESeq(ddsHTSeq)
ndata=as.data.frame(counts(dds,normalized=TRUE))
colnames(ndata)=colnames(x)

rn=rownames(ndata)
ensID=apply(array(as.character(rn)),1,function(z) unlist(strsplit(z, "\\|"))[1])
gene=apply(array(as.character(rn)),1,function(z) unlist(strsplit(z, "\\|"))[2])
mydata=cbind(ensID,gene,ndata)

write.table(mydata,file="DESeq2_normalized_counts.txt",sep="\t",col.names=NA)
#png("HistDesq2normFilter.png")
df.m <- melt(as.data.frame(ndata))
print(ggplot(df.m) + geom_density(aes(x = value, colour = variable)) + labs(x = NULL) + theme(legend.position='right') + scale_x_log10())
# dev.off() 
#
## transformation
rld <- rlogTransformation(dds, blind=TRUE)
rldm=assay(rld)
colnames(rldm)=colnames(x)

mydata=cbind(ensID,gene,rldm)
write.table(mydata,file="DESeq2_normalized_rld.txt",sep="\t",col.names=NA)

```
```{r, echo=FALSE, warning=FALSE,message=FALSE,include=FALSE}
png("DESeq2_PCA.png")
print(plotPCA(rld, intgroup=c("condition")))
## dev.copy(png,"DESeq2_PCA.png")
dev.off()
png("DESeq2_PCA_details.png")
# print(plotPCA(rld, intgroup=c("condition","fileName")))
print(plotPCA(rld, intgroup=c("condition","label")))
## dev.copy(png,"DESeq2_PCA_details.png")
dev.off()

```
## **Boxplot and PCA**

```{r, echo=FALSE,webgl=TRUE,message=FALSE}

## Boxplot and PCA
df.m <- melt(as.data.frame(rldm))

#dev.off()
par(mar=c(10,7,1,1))
boxplot(value~variable,las=2,data=df.m,main="Normalized Signal",
        ylab="Counts",col=as.numeric(as.factor(sampleinfo$condition)))

edf=rldm
tedf= t(edf)
pca=prcomp(tedf,scale.=T)
tedf1 = data.frame(tedf)
Phenotype=sampleinfo$condition
cell_rep=sampleinfo$label
tedf1$group = as.factor(Phenotype)

# plot(pca,type="lines")  #Decide how many PC's are relevant for plotting
  #pca$x[,1:3]  #look at first 3 PC's

plot3d(pca$x[,1:3],col = as.integer(tedf1$group),type="s",size=2)
group.v<-as.vector(cell_rep)
text3d(pca$x, pca$y, pca$z, group.v, cex=1.0, adj = 1.2)
# rgl.postscript("pca3d_DESeq2.pdf","pdf")

```
## **Similarity Heatmap**

```{r, echo=FALSE, warning=FALSE,message=FALSE,include=FALSE}
## clustering / heatmap
hmcol <- colorRampPalette(brewer.pal(9, "GnBu"))(100)
distsRL <- dist(t(assay(rld)))
mat <- as.matrix(distsRL)
rownames(mat) <- colnames(mat) <- with(colData(dds),paste(condition,sampleFiles , sep=" : "))
#if you just want the conditions use this line : rownames(mat) <- colnames(mat) <- with(colData(dds),condition)
png("DESeq2_heatmaps_samplebysample.png")
heatmap.2(mat, trace="none", col = rev(hmcol), margin=c(16, 16))
## dev.copy(png,"DESeq2_heatmaps_samplebysample.png")
dev.off()
```

```{r, echo=FALSE,warning=FALSE,message=FALSE}
d=Dist(tedf,method="pearson",diag=TRUE)
m=as.matrix(d)

new.palette=colorRampPalette(c("black","red","yellow","white"),space="rgb")
#levelplot(m[1:ncol(m),ncol(m):1],col.regions=new.palette(20))
heatmap(m,symm=TRUE,col=new.palette(20))

```

## **Differentially expressed genes (DEG) for all contrasts: Volcano plots, Pathways and DEG tables**

```{r, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}

## differentially expressed genes -----
contras=unlist(strsplit(CONTRASTS, split=" "))        
cat(contras,"\t",length(contras),"\t",contras[1],"\t",contras[2],"\n",file="readcontra.txt")
mfc=c()
mpval=c()
cons=c()
l <- htmltools::tagList()
pup = htmltools::tagList()
pdw = htmltools::tagList()

#if ((grepl("hg",SPECIES)||grepl("mm",SPECIES))&&(grepl("genes",params$dtype)))
#{
#system(paste0("cd ",getwd(),"; perl ../Scripts/gtf2bed.pl ",params$gtffile," |sort -k1,1 -k2,2n > karyobed.bed"))
#system(paste0("cd ",getwd(),"; cut -f1 karyobed.bed|uniq > chrs.txt; while read a ;do cat karyobed.bed | awk -F \"\\t\" -v a=$a \'{if ($1==a) {print}}\' > karyobed.${a}.bed;done < chrs.txt"))
#}

for(i in seq(1, length(contras), by = 2))
{{
res<-results(dds,contrast=c("condition",as.character(contras[i]),as.character(contras[i+1])))
## res<-res[order(res$padj),] for now
res1=as.data.frame(res)
# mfc=cbind(mfc,res1$log2FoldChange)
# mpval=cbind(mpval,res1$pvalue)
cons=c(cons,paste(contras[i],"-",contras[i+1],sep=""))

restmp=res1
restmp$FoldChange <- ifelse(restmp$log2FoldChange<0, -1/(2^restmp$log2FoldChange), 2^restmp$log2FoldChange)
mfc=cbind(mfc,restmp$FoldChange)
mpval=cbind(mpval,restmp$pvalue)
final=restmp
x=rownames(restmp)
ensID=apply(array(as.character(x)),1,function(z) unlist(strsplit(z, "\\|"))[1])
gene=apply(array(as.character(x)),1,function(z) unlist(strsplit(z, "\\|"))[2])
restmp=cbind(ensID,gene,restmp)
write.table(restmp,file=paste("DESeq2_deg_",contras[i],"_vs_",contras[i+1],".txt",sep=""),sep="\t",col.names=NA) 

## pathway enrichment --------------------------------------------
if (SPECIES=="hg19" | SPECIES=="mm10" | SPECIES=="hg38")
{
iup=which(final$pvalue<0.05 & final$FoldChange >=0)
idw=which(final$pvalue<0.05 & final$FoldChange <0)
fin.up=final[iup,]
if (length(iup) > 500)
  {
   fin.up=fin.up[order(fin.up$pvalue),]
   fin.up=fin.up[1:500,]
  }
x2=rownames(fin.up)
gup=apply(array(as.character(x2)),1,function(z) unlist(strsplit(z, "\\|"))[2])

fin.dw=final[idw,]
if (length(idw) > 500)
  {
   fin.dw=fin.dw[order(fin.dw$pvalue),]
   fin.dw=fin.dw[1:500,]
  }
x2=rownames(fin.dw)
gdw=apply(array(as.character(x2)),1,function(z) unlist(strsplit(z, "\\|"))[2])

extension=paste0("_",contras[i],"_vs_",contras[i+1],".txt")
if (SPECIES=="hg19" | SPECIES=="hg38")
{
write.table(gup,paste0("DESeq2_gene_up_path",extension),quote=F,row.names=F,col.names=F)
write.table(gdw,paste0("DESeq2_gene_dw_path",extension),quote=F,row.names=F,col.names=F)
}
else
{
write.table(gup,"DESeq2_temp_up.txt",quote=F,row.names=F,col.names=F)
write.table(gdw,"DESeq2_temp_dw.txt",quote=F,row.names=F,col.names=F)
system(paste0("cat DESeq2_temp_up.txt | grep -v \"^NA\" | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/m2h | grep -v XXXX | cut -f2 -d\" \"  >DESeq2_gene_up_path",extension))
system(paste0("cat DESeq2_temp_dw.txt | grep -v \"^NA\" | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/m2h | grep -v XXXX | cut -f2 -d\" \"  >DESeq2_gene_dw_path",extension))
}
# extension=paste0("_",contras[i],"_vs_",contras[i+1],".txt")
system(paste0("cat DESeq2_gene_up_path",extension," |sort | uniq | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/l2p >DESeq2_res_path_up",extension))
system(paste0("cat DESeq2_gene_dw_path",extension," |sort | uniq | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/l2p >DESeq2_res_path_dw",extension))

mup=read.delim(paste0("DESeq2_res_path_up",extension),header=F)
mdw=read.delim(paste0("DESeq2_res_path_dw",extension),header=F)
colnames(mup)=c("pval","fdr","ratio","nb.hits","nb.genes.path","nb.user.genes","tot.back.genes","path_id","source","description","type","gene.list")
colnames(mdw)=c("pval","fdr","ratio","nb.hits","nb.genes.path","nb.user.genes","tot.back.genes","path_id","source","description","type","gene.list")

#mup=mup[which(mup[,1]<0.05),]
#mdw=mdw[which(mdw[,1]<0.05),]

mup=mup[which(mup[,1]<0.05 & mup[,4]>=5),]
mdw=mdw[which(mdw[,1]<0.05 & mdw[,4]>=5),]

mup=mup[order(mup[,1]),]
mdw=mdw[order(mdw[,1]),]
mup=mup[,c(8,9,10,11,1:7,12)]
mdw=mdw[,c(8,9,10,11,1:7,12)]


pup[[i]] <- DT::datatable(mup, caption=htmltools::tags$caption(paste0("Pathways for top 500 up-regulated genes ",contras[i],"_vs_",contras[i+1]) ,style="color:red; font-size: 18px" ),rownames=F)
pdw[[i]] <- DT::datatable(mdw, caption=htmltools::tags$caption(paste0("Pathways for top 500 down-regulated genes ",contras[i],"_vs_",contras[i+1]) ,style="color:green; font-size: 18px" ),rownames=F )

}
## ------------------------------------------------------------------


x=res1$log2FoldChange[which(!is.na(res1$log2FoldChange))] 
png(paste("MAplot_",contras[i],"_vs_",contras[i+1],".png",sep=""))
plotMA(res,ylim=range(x),main=paste("MAplot_",contras[i],"_vs_",contras[i+1],sep=""))
##dev.copy(png,paste("DESeq2_MAplot_",contras[i],"_vs_",contras[i+1],".png",sep=""))
dev.off()
# Volcano

log_FC=res1$log2FoldChange
log_pval=-log10(res1$pvalue)
Significant=rep("NotSignificant",length(log_FC))
Significant[which(res1$pvalue<0.05 & abs(res1$log2FoldChange)>=1)]="Significant&LogFoldChange"
Significant[which(res1$pvalue<0.05 & abs(res1$log2FoldChange)<1)]="Significant"
Significant[which(res1$pvalue>=0.05 & abs(res1$log2FoldChange)>=1)]="LogFoldChange"
gene=rownames(res1)
volcano_data=as.data.frame(cbind(gene,log_FC,log_pval,Significant))
l[[i]] <- plot_ly(data = volcano_data, x = log_FC, y = log_pval, text = gene, mode = "markers", color = Significant) %>% layout(title =paste("Volcano Plot for ",contras[i],"_vs_",contras[i+1],sep=""),xaxis=list(title="Fold Change",range =c(-5,5),tickvals=c(-5,-4,-3,-2,-1,0,1,2,3,4,5),ticktext=c('-32','-16','-8','-4','-2','1','2','4','8','16','32')),yaxis=list(title="-Log10 pvalue",range =c(0,10)))

#l[[i]] <- plot_ly(data = volcano_data, x = log_FC, y = log_pval, text = gene, mode = "markers", color = Significant) %>% layout(title =paste("Volcano Plot for ",contras[i],"_vs_",contras[i+1],sep=""),xaxis=list(title="Log2_FC",range =c(-10,10)),yaxis=list(title="-Log10_pvalue",range=c(0,10)))
#  %>% layout(annotations = a)

if ((grepl("hg",SPECIES)||grepl("mm",SPECIES))&&(grepl("genes",params$dtype)))
{
system(paste0("cd ",getwd(),"; python ../Scripts/FC2karyo.py DESeq2_deg_",contras[i],"_vs_",contras[i+1],".txt 2 4 ",SPECIES," ",KARYOBEDS," > DESeq2_deg_",contras[i],"_vs_",contras[i+1],".input.karyo"))
system(paste0("cd ",getwd(),"; python ../Scripts/karyoplot.py DESeq2_deg_",contras[i],"_vs_",contras[i+1],".input.karyo ",SPECIES))
ki1=paste0("DESeq2_deg_",contras[i],"_vs_",contras[i+1],".input.karyo1.png")
ki2=paste0("DESeq2_deg_",contras[i],"_vs_",contras[i+1],".input.karyo2.png")
cat("\n###KARYO PLOT",contras[i],"_vs_",contras[i+1],"\n")
cat("![](",ki1,")")
cat("\n")
cat("![](",ki2,")")
cat("\n\n")
}

}}

colnames(mfc)=paste(cons,"FC",sep="_")
colnames(mpval)=paste(cons,"pval",sep="_")
res=as.data.frame(cbind(mfc,mpval))
rownames(res)=rownames(res1)
write.table(res,file="DESeq2_res_all_contrasts.txt",sep="\t",col.names=NA)

l

pup

pdw

DT::datatable(res, caption=htmltools::tags$caption("DEG for all contrasts",style="color:blue; font-size: 18px" ))
```

```{r, echo=TRUE, message=FALSE, warning=FALSE}

sessionInfo()

```

