---
title: "DiffBind: CCBR ChIP-seq pipeline"
output: 
    html_document:
        toc: true
        toc_depth: 2
params:
    csvfile: samplesheet.csv
    contrasts: "group1_vs_group2"
    peakcaller: "macs"
    projectID: "<projectID>"
    projectDesc: "<desc>"
---

```{r, include=FALSE, warning=FALSE, message=FALSE}
## grab args
projectID <- params$projectID
projectDesc <- params$projectDesc
dateandtime<-format(Sys.time(), "%a %b %d %Y - %X")

csvfile <- params$csvfile
contrasts <- params$contrasts
peakcaller <- params$peakcaller
```

### **Project:**
####    *`r projectID`*
### **Description:** 
####    *`r projectDesc`*
### **Groups being compared:**
####    *`r contrasts`*
### **Peak sources:**
####    *`r peakcaller`*
### **Report generated:** 
####    *`r dateandtime`* 

```{r setup, echo=FALSE, warning=FALSE,message=FALSE}
knitr::opts_chunk$set(echo = TRUE)
suppressMessages(library(DT))
suppressMessages(library(DiffBind))
```

<br/>

## Read in sample sheet information and peak information
```{r samples, echo=FALSE, warning=FALSE,message=FALSE} 
samples <- dba(sampleSheet=csvfile)
print(samples)
```

<br/>

## Plot raw information about the peaks
### Correlation heatmap: Only peaks
```{r heatmap1, echo=FALSE, warning=FALSE,message=FALSE,out.width = "80%",fig.align="center"}
try(plot(samples,main=""),silent=TRUE)
```

### PCA: Only peaks
```{r PCA1, echo=FALSE, warning=FALSE,message=FALSE,fig.height=5,fig.width=5,fig.align="center",fig.caption="PCA:\nOnlyPeaks"}
try(dba.plotPCA(samples,DBA_CONDITION),silent=TRUE)
```

### Overlapping peak counts
```{r Venn, echo=FALSE, warning=FALSE,message=FALSE,fig.align="center"}
if (nrow(samples$samples) < 5) {
   dba.plotVenn(samples,1:nrow(samples$samples)) 
} else {
   dba.plotVenn(samples,samples$masks[[3]])
   dba.plotVenn(samples,samples$masks[[4]])
   dba.plotVenn(samples,samples$masks$Replicate.1)
}
```

```{r peaksORsummits, echo=F}
if ( grepl("narrow",samples$samples$Peaks[1]) ) {
   summits <- TRUE
   print ("Narrow peak calling tool.")
   print ("Differential peaks are 250bp upstream and downstream of the summits.")
} else if ( grepl("broad",samples$samples$Peaks[1]) ) {
  summits <- FALSE
  print ("Broad peak calling tool.")
  print ("Differential peaks are consensus peaks.")
} else {
  summits <- FALSE
  print ("Indeterminate peak calling tool.")
  print ("Differential peaks are consensus peaks.")
}
```

## Read in bam file information under all peaks found in at least two samples
```{r DBcount, echo=FALSE, warning=FALSE,message=FALSE}
if (summits == TRUE) {
	DBdataCounts <- dba.count(samples, summits=250)
} else {
	DBdataCounts <- dba.count(samples)
}
print(DBdataCounts)
```

<br/>

## Plot raw information about all analyzed peaks
### Correlation heatmap: Peaks and reads
```{r heatmap2, echo=FALSE, warning=FALSE,message=FALSE,out.width = "80%",fig.align="center"}
try(plot(DBdataCounts, main=""),silent=TRUE)
```

### Heatmap: Average signal across each peak
```{r heatmap3, echo=FALSE, warning=FALSE,message=FALSE,out.width = "80%",fig.align="center"}
try(dba.plotHeatmap(DBdataCounts,correlations=FALSE),silent=TRUE)
```

### PCA: Peaks and reads
```{r PCA2, echo=FALSE, warning=FALSE,message=FALSE,fig.height=5,fig.width=5,fig.align="center"}
try(dba.plotPCA(DBdataCounts,DBA_CONDITION),silent=TRUE)
```

## Associate individual samples with the different contrasts
```{r contrast, echo=FALSE, warning=FALSE,message=FALSE}
DBdatacontrast <- dba.contrast(DBdataCounts, minMembers=2, categories = DBA_CONDITION)
print(DBdatacontrast)
```

<br/>

## Call differential peaks using Deseq2 and EdgeR
```{r analyze, echo=FALSE, warning=FALSE,message=FALSE}
DBAnalysisDeseq2 <- dba.analyze(DBdatacontrast, method = DBA_DESEQ2)
DBAnalysisEdgeR <- dba.analyze(DBdatacontrast, method = DBA_EDGER)
```

```{r, echo=FALSE, warning=FALSE,message=FALSE}
DBReportDeseq2 <- dba.report(DBAnalysisDeseq2, method = DBA_DESEQ2)
DBReportEdgeR <- dba.report(DBAnalysisEdgeR, method = DBA_EDGER)
```

### PCA: DeSeq2
```{r PCA3, echo=FALSE, warning=FALSE,message=FALSE,fig.height=5,fig.width=5,fig.align="center"}
try(dba.plotPCA(DBAnalysisDeseq2, contrast=1, method= DBA_DESEQ2),silent=TRUE)
```

### PCA: EdgeR
```{r PCA4, echo=FALSE, warning=FALSE,message=FALSE,fig.height=5,fig.width=5,fig.align="center"}
try(dba.plotPCA(DBAnalysisEdgeR, contrast=1, method = DBA_EDGER),silent=TRUE)
```

### MANorm: (left) Deseq2, (right) EdgeR
```{r MA, echo=FALSE, warning=FALSE,message=FALSE,fig.width=10,fig.height=4,fig.align="center"}
par(mfcol=c(1,2))
try(dba.plotMA(DBAnalysisDeseq2, method = DBA_DESEQ2),silent=TRUE)
try(dba.plotMA(DBAnalysisEdgeR, method = DBA_EDGER),silent=TRUE)
```

### Volcano plot: DeSeq2
```{r Volcano1, echo=FALSE, warning=FALSE,message=FALSE,out.width = "80%",fig.align="center"}
try(dba.plotVolcano(DBAnalysisDeseq2, method = DBA_DESEQ2),silent=TRUE)
```

### Volcano plot: EdgeR
```{r Volcano2, echo=FALSE, warning=FALSE,message=FALSE,out.width = "80%",fig.align="center"}
try(dba.plotVolcano(DBAnalysisEdgeR, method = DBA_EDGER),silent=TRUE)
```

### Boxplots: (left) Deseq2, (right) EdgeR
```{r BoxPlot, echo=FALSE, warning=FALSE,message=FALSE,fig.width=10,fig.height=4,fig.align="center"}
par(mfcol=c(1,2))
if (length(DBReportDeseq2) > 0) {
   try(dba.plotBox(DBAnalysisDeseq2, method = DBA_DESEQ2),silent=TRUE)
} else {
   plot(0,type='n',axes=FALSE,ann=FALSE)
}
try(dba.plotBox(DBAnalysisEdgeR, method = DBA_EDGER),silent=TRUE)
```

## Differentially bound peaks: Deseq2 output
```{r Deseq2Report, echo=FALSE, warning=FALSE,message=FALSE}
outfile <- paste0(contrasts, "-", peakcaller, "_Diffbind_Deseq2.txt")
write.table(DBReportDeseq2, outfile, quote=F, sep="\t", row.names=F)
DT::datatable(data.frame(DBReportDeseq2), rownames=F)
```

## Differentially bound peaks: EdgeR output
```{r EdgeRReport, echo=FALSE, warning=FALSE,message=FALSE}
outfile <- paste0(contrasts, "-", peakcaller,"_Diffbind_EdgeR.txt")
write.table(DBReportEdgeR, outfile, quote=F, sep="\t", row.names=F)
DT::datatable(data.frame(DBReportEdgeR), rownames=F)
```

## R tool version information
```{r Info, echo=FALSE, message=FALSE, warning=FALSE}
sessionInfo()
```