---
title: "EdgeR results"
author: "CCBR RNAseq pipeline"
output: html_document
params:
  folder: result
  sampleinfo: sampletable.txt
  data: RawCountFile_genes_filtered.txt
  contrasts: "GN GP"
  species: hg19
  projectId: "<projectId>"
  projectDesc: "<desc>"
  gtffile: "gencode.v19.annotation.gtf"
  dtype: "<count type>"
  karyobeds: "path/to/karyobedfiles"
---

```{r, include=FALSE, warning=FALSE, message=FALSE}
countMethod <- "subread"
if(grepl("rsem",params$data)){
  countMethod <- "rsem"
}
projectId<-params$projectId
projectDesc<-params$projectDesc
dateandtime<-format(Sys.time(), "%a %b %d %Y - %X")
```
### **Project:**
####    *`r projectId`*
### **Description:** 
####    *`r projectDesc`*
###  **Count method:** 
####    *`r countMethod`* 
### **Report generated:** 
####    *`r dateandtime`* 

## **Histogram after normalization**

```{r setup, echo=FALSE, warning=FALSE,message=FALSE}

library(rgl)
library(knitr)
library('edgeR')
library('statmod')
library('RColorBrewer') 
library('gplots')
library('reshape') 
library('ggplot2')
library('ggfortify')
library(amap)
library(DT)
library(plotly)
library('geneplotter')

knit_hooks$set(rgl = function(before, options, envir) {
  if (!before) {
    ## after a chunk has been evaluated
    if (rgl.cur() == 0) return()  # no active device
    name = paste(options$fig.path, options$label, sep = '')
    rgl.snapshot(paste(name, '.png', sep = ''), fmt = 'png')
    return(paste('\\includegraphics{', name, '}\n', sep = ''))
  }
})

knit_hooks$set(webgl = hook_webgl)
```

```{r, echo=FALSE, warning=FALSE,message=FALSE}

## grab args & normalization

DIR <- params$folder
FILE1 <- params$sampleinfo
FILE2 <- params$data
CONTRASTS <- params$contrasts
SPECIES <-params$species
KARYOBEDS <-params$karyobeds
# ANNOTATE <- args[4]

#
## setwd(DIR)
# read files
sampleinfo=read.delim(FILE1)
x = read.delim(FILE2,row.names=1)
colnames(x)=as.character(sampleinfo[,4])
# sampleFiles=as.character(sampleinfo[,2])
## read annotation file
## ann=read.delim(ANNOTATE)
# DGElist object --------------------------------------------------------------
condition = as.factor(sampleinfo$condition)
y = DGEList(counts=x,group=condition)
design=model.matrix(~0+condition)
contras=unlist(strsplit(CONTRASTS, split=" "))
cat(contras,"\t",length(contras),"\t",contras[1],"\t",contras[2],"\n",file="readcontra-edgeR.txt")
cons=c()
for(i in seq(1, length(contras), by = 2))
{{
cons=c(cons,paste(contras[i],"-",contras[i+1],sep=""))
}}

## Normalization TMM ------------------------------------------------------------
## method = =c("TMM","RLE","upperquartile","none")
y <- calcNormFactors(y,method="TMM")
# y$samples

#png("libdistrib.png")
barplot(y$samples$lib.size*1e-6,main="Library size distribution", names= strsplit(colnames(y$counts),".star.count.txt"), ylab="Library size (millions)",las=2,cex.names=0.8)
#dev.off()

## MDS plots ----------------------------------------------------------------------
# both pairewise (leading)

# png("MDS_bcv.png")
# print(y)
#plotMDS(y, method="bcv", , main="MDS plot bcv")
#dev.off()

# png("MDS_logFC.png")
plotMDS(y, method="logFC" , main="MDS plot logFC") ## plotMDS(y) default
# dev.off()
# plotMDS(y, method="logFC",gene.selection="common", main="MDS plot common")

## estimating common and tagwise dispersions -----------------------------------------
y <- estimateDisp(y, design, robust=TRUE)

# 
## plotting
# png("BCVplot.png")
 plotBCV(y,main="BCV plot")
#dev.off()

ylog2=cpm(y,log=TRUE,normalized.lib.sizes=TRUE,prior.count=2) # prior count like avelogcpm
ndata= cpm(y,log=FALSE,normalized.lib.sizes=TRUE)*1e6
## save it
rn=rownames(ylog2)
ensID=apply(array(as.character(rn)),1,function(z) unlist(strsplit(z, "\\|"))[1])
gene=apply(array(as.character(rn)),1,function(z) unlist(strsplit(z, "\\|"))[2])
mydata=cbind(ensID,gene,ylog2)

write.table(mydata,file="edgeR_normalized_counts_log.txt",sep="\t",col.names=NA)
mydata=cbind(ensID,gene,ndata)

write.table(mydata,file="edgeR_normalized_counts.txt",sep="\t",col.names=NA)
## png("HistEdgeRnormFilter.png")
df.m <- melt(as.data.frame(ndata))
print(ggplot(df.m) + geom_density(aes(x = value, colour = variable)) + labs(x = NULL) + theme(legend.position='right') + scale_x_log10() + ggtitle("Normalized Counts"))

```
```{r, echo=FALSE, warning=FALSE,message=FALSE,include=FALSE}
## dev.off()
#pr2=prcomp(t(ylog2))
#dd=cbind(t(ylog2),condition=as.character(condition))

#png("edgeR_prcomp.png")
#biplot(pr2)
#plot(pr2$x[,1],pr2$x[,2],col="red", main="PCA plot using prcomp and Logcpm data")
#text(pr2$x[,1],pr2$x[,2], labels=colnames(ylog2), cex=0.7, pos=4)
#dev.off()

#png("edgeR_pca.png")
#autoplot(pr2,data=dd, colour = 'condition')
#dev.off()

```
## **Boxplot and PCA**

```{r, echo=FALSE,webgl=TRUE,message=FALSE}

## Boxplot and PCA
df.m <- melt(as.data.frame(ylog2))

#dev.off()
par(mar=c(10,7,1,1))
boxplot(value~variable,las=2,data=df.m,main="Normalized Signal",
        ylab="Counts",col=as.numeric(as.factor(sampleinfo$condition)))

edf=ylog2
tedf= t(edf)
pca=prcomp(tedf,scale.=T)
tedf1 = data.frame(tedf)
Phenotype=sampleinfo$condition
cell_rep=sampleinfo$label
tedf1$group = as.factor(Phenotype)

# plot(pca,type="lines")  #Decide how many PC's are relevant for plotting
  #pca$x[,1:3]  #look at first 3 PC's

plot3d(pca$x[,1:3],col = as.integer(tedf1$group),type="s",size=2)
group.v<-as.vector(cell_rep)
text3d(pca$x, pca$y, pca$z, group.v, cex=1.0, adj = 1.2)
# rgl.postscript("pca3d_edgeR.pdf","pdf")

```

## **Similarity Heatmap**
```{r, echo=FALSE,warning=FALSE,message=FALSE,include=FALSE}

## clustering / heatmap
hmcol <- colorRampPalette(brewer.pal(9, "GnBu"))(100)
distylog2=dist(t(ylog2))
mat = as.matrix(distylog2)
# rownames(mat) <- colnames(mat)
#png("edgeR_heatmaps_samplebysample.png")
heatmap.2(mat, trace="none", col = rev(hmcol), margin=c(16, 16))
## dev.copy(png,"edgeR_heatmaps_samplebysample.png")
dev.off()

```


```{r, echo=FALSE,warning=FALSE,message=FALSE}

d=Dist(tedf,method="pearson",diag=TRUE)
m=as.matrix(d)

new.palette=colorRampPalette(c("black","red","yellow","white"),space="rgb")
#levelplot(m[1:ncol(m),ncol(m):1],col.regions=new.palette(20))
heatmap(m,symm=TRUE,col=new.palette(20))

```

## **Differentially expressed genes (DEG) for all contrasts: Volcano plots, Pathways and DEG tables**


```{r, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}

## differentially expressed genes -----
contras=unlist(strsplit(CONTRASTS, split=" "))        

l <- htmltools::tagList()


nb=length(contras)/2

for(i in 1:nb)
{{
  filename = paste(contras[i],"_vs_",contras[i+1], ".gene.EBSeq", sep="")
  res = read.delim(filename, header=T)

  
  log_FC=log2(res$RealFC)
  log_pval=res$PPDE
  Significant=rep("NotSignificant",length(log_FC))
  Significant[which(res$PPDE<0.05 & abs(log2(res$RealFC))>=1)]="Significant&LogFoldChange"
  Significant[which(res$PPDE<0.05 & abs(log2(res$RealFC))<1)]="Significant"
  Significant[which(res$PPDE>=0.05 & abs(log2(res$RealFC))>=1)]="LogFoldChange"
  gene=rownames(res)
  volcano_data=as.data.frame(cbind(gene,log_FC,log_pval,Significant))

  l <- plot_ly(data = volcano_data, x = log_FC, y = log_pval, text = gene, mode = "markers", color = Significant) %>% layout(title =paste("Volcano Plot for ",contras[f],"_vs_",contras[f+1],sep=""),xaxis=list(title="Fold Change",range =c(-5,5),tickvals=c(-5,-4,-3,-2,-1,0,1,2,3,4,5),ticktext=c('-32','-16','-8','-4','-2','1','2','4','8','16','32')),yaxis=list(title="-Log10 pvalue",range =c(0,1)))

  l
  
  res = res[which(res$PPDE>0.95 & abs(log2(res$RealFC))>=1),]
  
  DT::datatable(res, caption=htmltools::tags$caption(paste("DEG table for ", contras[i],"_vs_",contras[i+1]," style='color:blue; font-size: 18px'", sep="")))


}}


```

```{r, echo=TRUE, message=FALSE, warning=FALSE}

sessionInfo()

```


