---
title: "EdgeR results"
author: "CCBR RNAseq pipeline"
output: html_document
params:
  folder: result
  sampleinfo: sampletable.txt
  data: RawCountFile_genes_filtered.txt
  contrasts: "GN GP"
  species: hg19
  projectId: "<projectId>"
  projectDesc: "<desc>"
  gtffile: "gencode.v19.annotation.gtf"
  dtype: "<count type>"
---

```{r, include=FALSE, warning=FALSE, message=FALSE}
countMethod <- "subread"
if(grepl("rsem",params$data)){
  countMethod <- "rsem"
}
projectId<-params$projectId
projectDesc<-params$projectDesc
dateandtime<-format(Sys.time(), "%a %b %d %Y - %X")
```
### **Project:**
####    *`r projectId`*
### **Description:** 
####    *`r projectDesc`*
###  **Count method:** 
####    *`r countMethod`* 
### **Report generated:** 
####    *`r dateandtime`* 

## **Histogram after normalization**

```{r setup, echo=FALSE, warning=FALSE,message=FALSE}

library(rgl)
library(knitr)
library('edgeR')
library('statmod')
library('RColorBrewer') 
library('gplots')
library('reshape') 
library('ggplot2')
library('ggfortify')
library(amap)
library(DT)
library(plotly)
library('geneplotter')

knit_hooks$set(rgl = function(before, options, envir) {
  if (!before) {
    ## after a chunk has been evaluated
    if (rgl.cur() == 0) return()  # no active device
    name = paste(options$fig.path, options$label, sep = '')
    rgl.snapshot(paste(name, '.png', sep = ''), fmt = 'png')
    return(paste('\\includegraphics{', name, '}\n', sep = ''))
  }
})

knit_hooks$set(webgl = hook_webgl)
```

```{r, echo=FALSE, warning=FALSE,message=FALSE}

## grab args & normalization

DIR <- params$folder
FILE1 <- params$sampleinfo
FILE2 <- params$data
CONTRASTS <- params$contrasts
SPECIES <-params$species
# ANNOTATE <- args[4]

#
## setwd(DIR)
# read files
sampleinfo=read.delim(FILE1)
x = read.delim(FILE2,row.names=1)
colnames(x)=as.character(sampleinfo[,4])
# sampleFiles=as.character(sampleinfo[,2])
## read annotation file
## ann=read.delim(ANNOTATE)
# DGElist object --------------------------------------------------------------
condition = as.factor(sampleinfo$condition)
y = DGEList(counts=x,group=condition)
## Normalization TMM ------------------------------------------------------------
## method = =c("TMM","RLE","upperquartile","none")
y <- calcNormFactors(y,method="TMM")
# y$samples

#png("libdistrib.png")
barplot(y$samples$lib.size*1e-6,main="Library size distribution", names= strsplit(colnames(y$counts),".star.count.txt"), ylab="Library size (millions)",las=2,cex.names=0.8)
#dev.off()

## MDS plots ----------------------------------------------------------------------
# both pairewise (leading)

# png("MDS_bcv.png")
# print(y)
#plotMDS(y, method="bcv", , main="MDS plot bcv")
#dev.off()

# png("MDS_logFC.png")
plotMDS(y, method="logFC" , main="MDS plot logFC") ## plotMDS(y) default
# dev.off()
# plotMDS(y, method="logFC",gene.selection="common", main="MDS plot common")

## estimating common and tagwise dispersions -----------------------------------------
y <- estimateCommonDisp(y)
y <- estimateTagwiseDisp(y) #default trend: moveingave
## plotting
# png("BCVplot.png")
 plotBCV(y,main="BCV plot")
#dev.off()

ylog2=cpm(y,log=TRUE,normalized.lib.sizes=TRUE,prior.count=2) # prior count like avelogcpm
ndata= cpm(y,log=FALSE,normalized.lib.sizes=TRUE)*1e6
## save it
write.table(ylog2,file="edgeR_normalized_counts_log.txt",sep="\t",col.names=NA)
write.table(ndata,file="edgeR_normalized_counts.txt",sep="\t",col.names=NA)
## png("HistEdgeRnormFilter.png")
df.m <- melt(as.data.frame(ndata))
print(ggplot(df.m) + geom_density(aes(x = value, colour = variable)) + labs(x = NULL) + theme(legend.position='right') + scale_x_log10() + ggtitle("Normalized Counts"))

```
```{r, echo=FALSE, warning=FALSE,message=FALSE,include=FALSE}
## dev.off()
pr2=prcomp(t(ylog2))
dd=cbind(t(ylog2),condition=as.character(condition))

png("edgeR_prcomp.png")
#biplot(pr2)
plot(pr2$x[,1],pr2$x[,2],col="red", main="PCA plot using prcomp and Logcpm data")
text(pr2$x[,1],pr2$x[,2], labels=colnames(ylog2), cex=0.7, pos=4)
dev.off()

png("edgeR_pca.png")
autoplot(pr2,data=dd, colour = 'condition')
dev.off()

```
## **Boxplot and PCA**

```{r, echo=FALSE,webgl=TRUE,message=FALSE}

## Boxplot and PCA
df.m <- melt(as.data.frame(ylog2))

#dev.off()
par(mar=c(10,7,1,1))
boxplot(value~variable,las=2,data=df.m,main="Normalized Signal",
        ylab="Counts",col=as.numeric(as.factor(sampleinfo$condition)))

edf=ylog2
tedf= t(edf)
pca=prcomp(tedf,scale.=T)
tedf1 = data.frame(tedf)
Phenotype=sampleinfo$condition
cell_rep=sampleinfo$label
tedf1$group = as.factor(Phenotype)

# plot(pca,type="lines")  #Decide how many PC's are relevant for plotting
  #pca$x[,1:3]  #look at first 3 PC's

plot3d(pca$x[,1:3],col = as.integer(tedf1$group),type="s",size=2)
group.v<-as.vector(cell_rep)
text3d(pca$x, pca$y, pca$z, group.v, cex=1.0, adj = 1.2)
# rgl.postscript("pca3d_edgeR.pdf","pdf")

```

## **Similarity Heatmap**
```{r, echo=FALSE,warning=FALSE,message=FALSE,include=FALSE}

## clustering / heatmap
hmcol <- colorRampPalette(brewer.pal(9, "GnBu"))(100)
distylog2=dist(t(ylog2))
mat = as.matrix(distylog2)
# rownames(mat) <- colnames(mat)
png("edgeR_heatmaps_samplebysample.png")
heatmap.2(mat, trace="none", col = rev(hmcol), margin=c(16, 16))
## dev.copy(png,"edgeR_heatmaps_samplebysample.png")
dev.off()

```


```{r, echo=FALSE,warning=FALSE,message=FALSE}

d=Dist(tedf,method="pearson",diag=TRUE)
m=as.matrix(d)

new.palette=colorRampPalette(c("black","red","yellow","white"),space="rgb")
#levelplot(m[1:ncol(m),ncol(m):1],col.regions=new.palette(20))
heatmap(m,symm=TRUE,col=new.palette(20))

```

## **Differentially expressed genes (DEG) for all contrasts: Volcano plots, Pathways and DEG tables**


```{r, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}

## differentially expressed genes -----
contras=unlist(strsplit(CONTRASTS, split=" "))        
cat(contras,"\t",length(contras),"\t",contras[1],"\t",contras[2],"\n",file="readcontra-edgeR.txt")
mfc=c()
mpval=c()
cons=c()
l <- htmltools::tagList()
pup = htmltools::tagList()
pdw = htmltools::tagList()

#if ((grepl("hg",SPECIES)||grepl("mm",SPECIES))&&(grepl("genes",params$dtype)))
#{
#system(paste0("cd ",getwd(),"; perl ../Scripts/gtf2bed.pl ",params$gtffile," |sort -k1,1 -k2,2n > karyobed.bed"))
#system(paste0("cd ",getwd(),"; cut -f1 karyobed.bed|uniq > chrs.txt; while read a ;do cat karyobed.bed | awk -F \"\\t\" -v a=$a \'{if ($1==a) {print}}\' > karyobed.${a}.bed;done < chrs.txt"))
#}

for(i in seq(1, length(contras), by = 2))
{{
deg<-exactTest(y,c(as.character(contras[i+1]),as.character(contras[i])))
# 
n=dim(y$counts)[1]
tt=topTags(deg, n=n,sort.by="none")
res1 = as.data.frame(tt)
#mfc=cbind(mfc,res1$logFC)
#mpval=cbind(mpval,res1$PValue)
cons=c(cons,paste(contras[i],"-",contras[i+1],sep=""))
#
## res1=cbind(Ensembl.Gene.ID=substr(rownames(res1),1,18),id.ver=rownames(res1),res1)
## final= merge(res1,ann,all.x=TRUE)
##final=final[order(final$FDR),]
## final=res1[order(res1$FDR),] for now 
final=res1
final$FC <- ifelse(final$logFC<0, -1/(2^final$logFC), 2^final$logFC)
mfc=cbind(mfc,final$FC)
mpval=cbind(mpval,final$PValue)

x=rownames(final)
ensID=apply(array(as.character(x)),1,function(z) unlist(strsplit(z, "\\|"))[1])
gene=apply(array(as.character(x)),1,function(z) unlist(strsplit(z, "\\|"))[2])
final=cbind(ensID,gene,final)
write.table(final,file=paste("EdgeR_deg_",contras[i],"_vs_",contras[i+1],".txt",sep=""),sep="\t",col.names=NA)

## pathway enrichment --------------------------------------------
if (SPECIES=="hg19" | SPECIES=="mm10")
{
#gup=gene[which(final$PValue<0.05 & final$FC >=1.5)]
#gdw=gene[which(final$PValue<0.05 & final$FC <=-1.5)]
iup=which(final$PValue<0.05 & final$FC >=0)
idw=which(final$PValue<0.05 & final$FC <0)
fin.up=final[iup,]
if (length(iup) > 500)
  {
   fin.up=fin.up[order(fin.up$PValue),]
   fin.up=fin.up[1:500,]
  }
x2=rownames(fin.up)
gup=apply(array(as.character(x2)),1,function(z) unlist(strsplit(z, "\\|"))[2])

fin.dw=final[idw,]
if (length(idw) > 500)
  {
   fin.dw=fin.dw[order(fin.dw$PValue),]
   fin.dw=fin.dw[1:500,]
  }
x2=rownames(fin.dw)
gdw=apply(array(as.character(x2)),1,function(z) unlist(strsplit(z, "\\|"))[2])

extension=paste0("_",contras[i],"_vs_",contras[i+1],".txt")
if (SPECIES=="hg19")
{
write.table(gup,paste0("edgeR_gene_up_path",extension),quote=F,row.names=F,col.names=F)
write.table(gdw,paste0("edgeR_gene_dw_path",extension),quote=F,row.names=F,col.names=F)
}
else
{
write.table(gup,"temp_up.txt",quote=F,row.names=F,col.names=F)
write.table(gdw,"temp_dw.txt",quote=F,row.names=F,col.names=F)
system(paste0("cat temp_up.txt | grep -v \"^NA\" | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/m2h | grep -v XXXX | cut -f2 -d\" \"  >edgeR_gene_up_path",extension))
system(paste0("cat temp_dw.txt | grep -v \"^NA\" | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/m2h | grep -v XXXX | cut -f2 -d\" \"  >edgeR_gene_dw_path",extension))
}
# extension=paste0("_",contras[i],"_vs_",contras[i+1],".txt")
system(paste0("cat edgeR_gene_up_path",extension," |sort | uniq | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/l2p >edgeR_res_path_up",extension))
system(paste0("cat edgeR_gene_dw_path",extension," |sort | uniq | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/l2p >edgeR_res_path_dw",extension))

mup=read.delim(paste0("edgeR_res_path_up",extension),header=F)
mdw=read.delim(paste0("edgeR_res_path_dw",extension),header=F)
colnames(mup)=c("pval","fdr","ratio","nb.hits","nb.genes.path","nb.user.genes","tot.back.genes","path_id","source","description","type","gene.list")
colnames(mdw)=c("pval","fdr","ratio","nb.hits","nb.genes.path","nb.user.genes","tot.back.genes","path_id","source","description","type","gene.list")

#mup=mup[which(mup[,1]<0.05),]
#mdw=mdw[which(mdw[,1]<0.05),]

mup=mup[which(mup[,1]<0.05 & mup[,4]>=5),]
mdw=mdw[which(mdw[,1]<0.05 & mdw[,4]>=5),]

mup=mup[order(mup[,1]),]
mdw=mdw[order(mdw[,1]),]
mup=mup[,c(8,9,10,11,1:7,12)]
mdw=mdw[,c(8,9,10,11,1:7,12)]


pup[[i]] <- DT::datatable(mup, caption=htmltools::tags$caption(paste0("Pathways for top 500 up-regulated genes ",contras[i],"_vs_",contras[i+1]) ,style="color:red; font-size: 18px" ),rownames=F)
pdw[[i]] <- DT::datatable(mdw, caption=htmltools::tags$caption(paste0("Pathways for top 500 down-regulated genes ",contras[i],"_vs_",contras[i+1]) ,style="color:green; font-size: 18px" ),rownames=F )

}
## ------------------------------------------------------------------

#  like MAplot
deg1sel <- decideTestsDGE(deg, p=0.05, adjust="BH")
detags <- rownames(y)[as.logical(deg1sel)]
png(paste("edgeR_Smearplot_",contras[i],"_vs_",contras[i+1],".png",sep=""))
plotSmear(deg, de.tags=detags,main= paste("Smearplot FDR<0.05 ",contras[i],"_vs_",contras[i+1],sep=""))
abline(h = c(-2, 2), col = "blue")
dev.off()
# volcano

log_FC=final$logFC
log_pval=-log10(final$PValue)
Significant=rep("NotSignificant",length(log_FC))
Significant[which(final$PValue<0.05 & abs(final$logFC)>=1)]="Significant&LogFoldChange"
Significant[which(final$PValue<0.05 & abs(final$logFC)<1)]="Significant"
Significant[which(final$PValue>=0.05 & abs(final$logFC)>=1)]="LogFoldChange"
gene=rownames(final)
volcano_data=as.data.frame(cbind(gene,log_FC,log_pval,Significant))
# l[[i]] <- plot_ly(data = volcano_data, x = log_FC, y = log_pval, text = gene, mode = "markers", color = Significant) %>% layout(title =paste("Volcano Plot for ",contras[i],"_vs_",contras[i+1],sep=""),xaxis=list(title="Log2_FC",range =c(-10,10)),yaxis=list(title="-Log10_pvalue",range =c(0,10)))
#  %>% layout(annotations = a)
l[[i]] <- plot_ly(data = volcano_data, x = log_FC, y = log_pval, text = gene, mode = "markers", color = Significant) %>% layout(title =paste("Volcano Plot for ",contras[i],"_vs_",contras[i+1],sep=""),xaxis=list(title="Fold Change",range =c(-5,5),tickvals=c(-5,-4,-3,-2,-1,0,1,2,3,4,5),ticktext=c('-32','-16','-8','-4','-2','1','2','4','8','16','32')),yaxis=list(title="-Log10 pvalue",range =c(0,10)))

if ((grepl("hg",SPECIES)||grepl("mm",SPECIES))&&(grepl("genes",params$dtype)))
{
system(paste0("cd ",getwd(),"; python ../Scripts/FC2karyo.py EdgeR_deg_",contras[i],"_vs_",contras[i+1],".txt 2 3 ",SPECIES," > EdgeR_deg_",contras[i],"_vs_",contras[i+1],".input.karyo"))
system(paste0("cd ",getwd(),"; python ../Scripts/karyoplot.py EdgeR_deg_",contras[i],"_vs_",contras[i+1],".input.karyo ",SPECIES))
ki1=paste0("EdgeR_deg_",contras[i],"_vs_",contras[i+1],".input.karyo1.png")
ki2=paste0("EdgeR_deg_",contras[i],"_vs_",contras[i+1],".input.karyo2.png")
cat("\n###KARYO PLOT",contras[i],"_vs_",contras[i+1],"\n")
cat("![](",ki1,")")
cat("\n")
cat("![](",ki2,")")
cat("\n\n")
}

}}

# add DEG  table
colnames(mfc)=paste(cons,"logFC",sep="_")
colnames(mpval)=paste(cons,"pval",sep="_")
res=as.data.frame(cbind(mfc,mpval))
rownames(res)=rownames(res1)
write.table(res,file="EdgeR_res_all_contrasts.txt",sep="\t",col.names=NA)

l

pup

pdw

DT::datatable(res, caption=htmltools::tags$caption("DEG for all contrasts",style="color:blue; font-size: 18px" ))
```

```{r, echo=TRUE, message=FALSE, warning=FALSE}

sessionInfo()

```


