---
title: "Limma Results"
author: "CCBR RNAseq pipeline"
output: html_document
params:
  folder: result
  sampleinfo: sampletable.txt
  data: RawCountFile_genes_filtered.txt
  contrasts: "GN GP"
  species: hg19
  projectId: "<projectId>"
  projectDesc: "<desc>"
  gtffile: "gencode.v19.annotation.gtf"
  dtype: "<count type>"
  karyobeds: "path/to/karyobedfiles"
  karyotext: "path/to/karyoplotertxt"
---

```{r params, include=FALSE, warning=FALSE, message=FALSE}
countMethod <- "subread"
if(grepl("RSEM",params$data)){
  countMethod <- "rsem"
}
projectId<-params$projectId
projectDesc<-params$projectDesc
dateandtime<-format(Sys.time(), "%a %b %d %Y - %X")
```
### **Project:**
####    *`r projectId`*
### **Description:** 
####    *`r projectDesc`*
###  **Count method:** 
####    *`r countMethod`* 
### **Report generated:** 
####    *`r dateandtime`* 
### **Session Info:** {.tabset .tabset-fade .tabset-pills}
#### Hide
```{r setup, echo=FALSE, warning=FALSE,message=FALSE}

suppressMessages(library(rgl))
suppressMessages(library(knitr))
suppressMessages(library('edgeR'))
suppressMessages(library('statmod'))
suppressMessages(library('RColorBrewer'))
suppressMessages(library(gplots))
suppressMessages(library('reshape'))
suppressMessages(library('ggplot2'))
suppressMessages(library('limma'))
suppressMessages(library('geneplotter'))
suppressMessages(library('ggfortify'))
suppressMessages(library(amap))
suppressMessages(library(DT))
suppressMessages(library(plotly))

#knitr::opts_chunk$set(fig.width=10, fig.height=8) 
knit_hooks$set(rgl = function(before, options, envir) {
  if (!before) {
    ## after a chunk has been evaluated
    if (rgl.cur() == 0) return()  # no active device
    name = paste(options$fig.path, options$label, sep = '')
    rgl.snapshot(paste(name, '.png', sep = ''), fmt = 'png')
    return(paste('\\includegraphics{', name, '}\n', sep = ''))
  }
})

knit_hooks$set(webgl = hook_webgl)
```

#### Show
```{r sessioninfo, echo=FALSE, warning=FALSE,message=FALSE}
sessionInfo()
```

## **Count Distribution** {.tabset}

### **Before Normalization**

```{r init, echo=FALSE, warning=FALSE, message=FALSE, webgl=TRUE,include=FALSE}

# Grab args 

DIR <- params$folder
FILE1 <- params$sampleinfo
FILE2 <- params$data
CONTRASTS <- params$contrasts
SPECIES <-params$species
KARYOBEDS <-params$karyobeds
KARYOTEXT <-params$karyotext

 
# Read in files
sampleinfo=read.delim(FILE1)
x = read.delim(FILE2,row.names=1)
colnames(x)=as.character(sampleinfo[,4])

myconditions = as.factor(sampleinfo$condition)
dge_before = DGEList(counts=x,group=myconditions)
dge_before <- calcNormFactors(dge_before,method="TMM")

Group <- factor(sampleinfo$condition)
design=model.matrix(~0+Group)
contras=unlist(strsplit(CONTRASTS, split=" "))  

cat(contras,"\t",length(contras),"\t",contras[1],"\t",contras[2],"\n",file="readcontralimma.txt")

cons=c()
for(i in seq(1, length(contras), by = 2))
{{
cons=c(cons,paste(contras[i],"-",contras[i+1],sep=""))
}}

png("limma_voom_meanVariance_Plot.png")
v1 <- voom(as.matrix(x),design,plot=TRUE,normalize="quantile")
dev.off()
```

```{r before_hist, echo=FALSE, warning=FALSE, message=FALSE, webgl=TRUE}
dge_before <- estimateCommonDisp(dge_before)
dge_before <- estimateTagwiseDisp(dge_before)
rawlog2= cpm(dge_before,log=TRUE,normalized.lib.sizes=FALSE,prior.count=0.5)

beforehist <- ggplotly(ggplot(melt(as.data.frame(rawlog2))) + geom_line(stat="density", aes(x = value,colour = variable)) + labs(x = NULL) + theme(legend.position='right') + scale_x_log10() + theme_bw())
beforehist
```

### Limma

```{r limma_hist, echo=FALSE, warning=FALSE, message=FALSE, webgl=TRUE}

sf = v1$E/log2((x/colSums(x))*1000000)

rn=rownames(v1$E)
ensID=apply(array(as.character(rn)),1,function(z) unlist(strsplit(z, "\\|"))[1])
gene=apply(array(as.character(rn)),1,function(z) unlist(strsplit(z, "\\|"))[2])
mydata=cbind(ensID,gene,v1$E)

sf=cbind(ensID,gene,sf)

write.table(sf,file="limma_voom_scaling_factors.txt",row.names=TRUE,col.names=NA,sep="\t",quote=FALSE)
write.table(mydata,file="limma_voom_normalized_data.txt",row.names=TRUE,col.names=NA,sep="\t",quote=FALSE)

#png("HistlimmavoomNormFilter.png")
df.n <- melt(as.data.frame(v1$E))
#print(ggplot(df.n) + geom_density(aes(x = value,colour = variable)) + labs(x = NULL) + theme(legend.position='right') + ggtitle("Normalized Counts"))
# dev.off()

limmahist <- ggplotly(ggplot(df.n) + geom_line(stat="density", aes(x = value,colour = variable)) + labs(x = NULL) + theme(legend.position='right') + scale_x_log10() + theme_bw())
limmahist
```
```{r limma_mds, echo=FALSE, warning=FALSE,message=FALSE,include=FALSE}
## MDS
png("limma_MDS.png")
## MDS <- plotMDS(v1,xlim=c(-5,5),ylim=c(-5,5),cex=1,pch=20)
plotMDS(v1,xlim=c(-5,5),ylim=c(-5,5),cex=1,pch=20)
## png("limma_MDS.png")
shortname=paste(substr(colnames(v1$E),1,22))
#text(MDS, labels=shortname, cex=0.5, pos=1)
# print(MDS)
## dev.copy(png, paste("limma_MDS.png"))
dev.off()

## 
```

## **Boxplot** {.tabset}

### **Before Normalization**

```{r before_hist, set-options, echo=FALSE,webgl=TRUE,message=FALSE}
before.dfm <- melt(as.data.frame(rawlog2))
par(mar=c(10,7,1,1))
boxplot(value~variable,las=2,data=before.dfm,
  ylab="Counts", col= c("#A6CEE3", "#B2DF8A")[as.numeric(as.factor(sampleinfo$condition))]) #col=as.numeric(as.factor(sampleinfo$condition)))
```

### Limma

```{r limma_hist, set-options, echo=FALSE,webgl=TRUE,message=FALSE}
#options(out.width = 160)

df.m <- melt(as.data.frame(v1$E))

par(mar=c(10,7,1,1))
boxplot(value~variable,las=2,data=df.m,
  ylab="Counts", col= c("#A6CEE3", "#B2DF8A")[as.numeric(as.factor(sampleinfo$condition))]) #col=as.numeric(as.factor(sampleinfo$condition)))
```


## **PCA Plot** {.tabset}

### **Before Normalization**

```{r before_pca, echo=FALSE,warning=FALSE,webgl=TRUE,message=FALSE,fig.width=9, fig.height=6}
# PCA for before norm
before.edf=rawlog2
before.tedf= t(before.edf)
before.pca=prcomp(before.tedf,scale.=T)
before.tedf1 = data.frame(before.tedf)
Phenotype=sampleinfo$condition
cell_rep=sampleinfo$label
before.tedf1$group = as.factor(Phenotype)
before.pc1 = round(before.pca$sdev[1]^2/sum(before.pca$sdev^2)*100,2)
before.pc2 = round(before.pca$sdev[2]^2/sum(before.pca$sdev^2)*100,2)
before.pc3 = round(before.pca$sdev[3]^2/sum(before.pca$sdev^2)*100,2)

pcafactor = as.factor(sampleinfo$condition)
col <- brewer.pal(nlevels(pcafactor), "Paired")

p <- plot_ly(as.data.frame(before.pca$x[,1:3]), x = ~PC1, y = ~PC2, z = ~PC3, color = pcafactor, colors = col, hoverinfo="text",
             hovertext = ~sampleinfo$label) %>%
  add_markers() %>%
  layout(scene = list(xaxis = list(title = paste0("PC1 (",before.pc1,"%)")),
                      yaxis = list(title = paste0("PC2 (",before.pc2,"%)")),
                      zaxis = list(title = paste0("PC3 (",before.pc3,"%)"))))
p
```

### Limma

```{r limma_pca, echo=FALSE,warning=FALSE,webgl=TRUE,message=FALSE,fig.width=9, fig.height=6}
edf=as.matrix(v1$E)
tedf= t(edf)
tedf=tedf[,apply(tedf,2,var)!= 0]
pca=prcomp(tedf,scale.=T)
tedf1 = data.frame(tedf)
Phenotype=sampleinfo$condition
cell_rep=sampleinfo$label
tedf1$group = as.factor(Phenotype)

pc1 = round(pca$sdev[1]^2/sum(pca$sdev^2)*100,2)
pc2 = round(pca$sdev[2]^2/sum(pca$sdev^2)*100,2)
pc3 = round(pca$sdev[3]^2/sum(pca$sdev^2)*100,2)

pcafactor = as.factor(sampleinfo$condition)

library(RColorBrewer)

col <- brewer.pal(nlevels(pcafactor), "Paired")

p <- plot_ly(as.data.frame(pca$x[,1:3]), x = ~PC1, y = ~PC2, z = ~PC3, color = pcafactor, colors = col, hoverinfo="text",
             hovertext = ~sampleinfo$label) %>%
  add_markers() %>%
  layout(title = "PCA plot", 
         scene = list(xaxis = list(title = paste0("PC1 (",pc1,"%)")),
                      yaxis = list(title = paste0("PC2 (",pc2,"%)")),
                      zaxis = list(title = paste0("PC3 (",pc3,"%)"))))

#saveWidget(as_widget(p), "Limma_PCA.html")

p
```

```{r heatmaps, echo=FALSE, warning=FALSE,message=FALSE,include=FALSE}

# PCA 2D
pr2=prcomp(tedf,scale.=T)
condition = factor(sampleinfo$condition)
dd=cbind(tedf,condition=as.character(condition))

png("limma_PCA_2D.png")
autoplot(pr2,data=dd, colour = 'condition')
dev.off()
```

## **Pearson Correlation Heatmap** {.tabset}

### **Before Normalization**

```{r before_heatmaps, echo=FALSE,warning=FALSE,message=FALSE}

d=Dist(before.tedf,method="pearson",diag=TRUE)
m=as.matrix(d)

new.palette=colorRampPalette(c("black","red","yellow","white"),space="rgb")
heatmap(m,symm=TRUE,col=new.palette(20))
```

### Limma

```{r limma_heatmaps, echo=FALSE,warning=FALSE,message=FALSE}

d=Dist(tedf,method="pearson",diag=TRUE)
m=as.matrix(d)

new.palette=colorRampPalette(c("black","red","yellow","white"),space="rgb")
#levelplot(m[1:ncol(m),ncol(m):1],col.regions=new.palette(20))
heatmap(m,symm=TRUE,col=new.palette(20))
```

## **Differentially Expression Analysis: `r params$contrasts`**
*** 
```{r deg, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}

# Differntially expressed genes
nb=length(contras)/2
colnames(design) <- levels(Group)
fit <- lmFit(v1,design)
contrast.matrix <- makeContrasts(contrasts=cons,levels=design)
fitb <- contrasts.fit(fit, contrast.matrix)
ebayes.fit=eBayes(fitb)

l <- htmltools::tagList()
v_fdr <- htmltools::tagList()
pup = htmltools::tagList()
pdw = htmltools::tagList()

mfc=c()
mpval=c()


for (i in 1:nb) {
  all.genes.con = topTable(ebayes.fit, coef = i, number=nrow(ebayes.fit), sort.by="none")
  
  all.genes.con$FC <- ifelse(all.genes.con$logFC<0, -1/(2^all.genes.con$logFC), 2^all.genes.con$logFC)
  final=all.genes.con
  mfc=cbind(mfc,final$FC)
  mpval=cbind(mpval,final$P.Value)
  
  ## Pathway enrichment analysis (Over-representation test ~ l2p)
  if (SPECIES=="hg19" | SPECIES=="mm10" | SPECIES=="mm10_M21" | SPECIES=="hg38" | SPECIES=="hg38_30") {
    
    iup=which(final$P.Value<0.05 & final$FC >=0)
    idw=which(final$P.Value<0.05 & final$FC <0)
    fin.up=final[iup,]

    if (length(iup) > 500) {
       fin.up=fin.up[order(fin.up$P.Value),]
       fin.up=fin.up[1:500,]
    }

    x2=rownames(fin.up)
    gup=apply(array(as.character(x2)),1,function(z) unlist(strsplit(z, "\\|"))[2])
    fin.dw=final[idw,]

    if (length(idw) > 500) {
       fin.dw=fin.dw[order(fin.dw$P.Value),]
       fin.dw=fin.dw[1:500,]
    }

    x2=rownames(fin.dw)
    gdw=apply(array(as.character(x2)),1,function(z) unlist(strsplit(z, "\\|"))[2])
    
    extension=paste0("_",cons[i],".txt")

    if (SPECIES=="hg19" | SPECIES=="hg38" | SPECIES=="hg38_30") {
      write.table(gup,paste0("limma_gene_up_path",extension),quote=F,row.names=F,col.names=F)
      write.table(gdw,paste0("limma_gene_dw_path",extension),quote=F,row.names=F,col.names=F)
    }
    else {
      write.table(gup,"limma_temp_up.txt",quote=F,row.names=F,col.names=F)
      write.table(gdw,"limma_temp_dw.txt",quote=F,row.names=F,col.names=F)
      system(paste0("cat limma_temp_up.txt | grep -v \"^NA\" | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/m2h | grep -v XXXX | cut -f2 -d\" \"  >limma_gene_up_path",extension))
      system(paste0("cat limma_temp_dw.txt | grep -v \"^NA\" | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/m2h | grep -v XXXX | cut -f2 -d\" \"  >limma_gene_dw_path",extension))
    }

    system(paste0("cat limma_gene_up_path",extension," |sort | uniq | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/l2p >limma_res_path_up",extension))
    system(paste0("cat limma_gene_dw_path",extension," |sort | uniq | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/l2p >limma_res_path_dw",extension))
    
    mup=read.delim(paste0("limma_res_path_up",extension),header=F)
    mdw=read.delim(paste0("limma_res_path_dw",extension),header=F)
    colnames(mup)=c("pval","fdr","ratio","nb.hits","nb.genes.path","nb.user.genes","tot.back.genes","path_id","source","description","type","gene.list")
    colnames(mdw)=c("pval","fdr","ratio","nb.hits","nb.genes.path","nb.user.genes","tot.back.genes","path_id","source","description","type","gene.list")
    
    mup=mup[which(mup[,1]<0.05 & mup[,4]>=5),]
    mdw=mdw[which(mdw[,1]<0.05 & mdw[,4]>=5),]
    
    mup=mup[order(mup[,1]),]
    mdw=mdw[order(mdw[,1]),]
    mup=mup[,c(8,9,10,11,1:7,12)]
    mdw=mdw[,c(8,9,10,11,1:7,12)]
    
    
    pup[[i]] <- DT::datatable(mup, extensions = 'Responsive', caption=htmltools::tags$caption(paste0("Pathways for top 500 up-regulated genes ",cons[i]) ,style="color:gray; font-size: 18px" ),rownames=F)
    pdw[[i]] <- DT::datatable(mdw, extensions = 'Responsive', caption=htmltools::tags$caption(paste0("Pathways for top 500 down-regulated genes ",cons[i]) ,style="color:gray; font-size: 18px" ),rownames=F )
    
  }
  
  # Volcano Plots 
  ## P-value and logFC

  log_FC=all.genes.con$logFC
  log_pval=-log10(all.genes.con$P.Value)
  Significant=rep("NotSignificant",length(log_FC))
  Significant[which(all.genes.con$P.Value<0.05 & abs(all.genes.con$logFC)>=1)]="Significant&LogFoldChange"
  Significant[which(all.genes.con$P.Value<0.05 & abs(all.genes.con$logFC)<1)]="Significant"
  Significant[which(all.genes.con$P.Value>=0.05 & abs(all.genes.con$logFC)>=1)]="LogFoldChange"
  gene=rownames(all.genes.con)
  volcano_data=as.data.frame(cbind(gene,log_FC,log_pval,Significant))
  l[[i]] <- plot_ly(data = volcano_data, x = log_FC, y = log_pval, text = gene, mode = "markers", color = Significant) %>% layout(title =paste("Volcano Plot for ",cons[i],sep=""),xaxis=list(title="Fold Change",range =c(-5,5),tickvals=c(-5,-4,-3,-2,-1,0,1,2,3,4,5),ticktext=c('-32','-16','-8','-4','-2','1','2','4','8','16','32')),yaxis=list(title="-Log10 pvalue",range =c(0,10)))
  
  # Volcano Plots 
  ## FDR and logFC

  log_FC=final$logFC
  log_pval=-log10(final$adj.P.Val)
  Significant=rep("NotSignificant",length(log_FC))
  Significant[which(final$adj.P.Val<0.05 &  abs(final$logFC)>=1)]="Significant&LogFoldChange"
  Significant[which(final$adj.P.Val<0.05 &  abs(final$logFC)<1)]="Significant"
  Significant[which(final$adj.P.Val>=0.05 & abs(final$logFC)>=1)]="LogFoldChange"
  gene=rownames(final)
  volcano_data=as.data.frame(cbind(gene,log_FC,log_pval,Significant))
  v_fdr[[i]] <- plot_ly(data = volcano_data, x = log_FC, y = log_pval, text = gene, mode = "markers", color = Significant) %>% layout(title =paste("Volcano Plot for ",cons[i],sep=""),xaxis=list(title="Fold Change",range =c(-5,5),tickvals=c(-5,-4,-3,-2,-1,0,1,2,3,4,5),ticktext=c('-32','-16','-8','-4','-2','1','2','4','8','16','32')),yaxis=list(title="-Log10 pvalue",range =c(0,10)))

  # MAplot 
  dataf=data.frame("m"=all.genes.con$AveExpr,"fc"=all.genes.con$logFC,"sig"=all.genes.con$adj.P.Val<0.05)
  png(paste(cons[i],"_MAplot_limma_v2.png",sep=""))
  plotMA(dataf,log="",main=cons[i],ylim=range(all.genes.con$logFC))
  dev.off()

  x=rownames(all.genes.con)
  ensID=apply(array(as.character(x)),1,function(z) unlist(strsplit(z, "\\|"))[1])
  gene=apply(array(as.character(x)),1,function(z) unlist(strsplit(z, "\\|"))[2])
  all.genes.con=cbind(ensID,gene,all.genes.con)
  write.table(all.genes.con,file=paste("limma_DEG_",cons[i],"_all_genes.txt",sep=""),sep="\t",col.names=NA)
  
  #Reformatted DEG File
  limmaout=read.table(paste("limma_DEG_",cons[i],"_all_genes.txt",sep=""), header = TRUE)
  limmaout=limmaout[,which(names(limmaout) %in% c("X","gene","logFC","P.Value"))]
  limmaout$fc=2^limmaout$logFC
  down_reg=limmaout$logFC<0
  limmaout$fc[down_reg]=-1/limmaout$fc[down_reg]
  limmaout=limmaout[,c("X","gene","fc","logFC","P.Value")]
  colnames(limmaout)=c("ensid_gene","gene","fc","log2fc","pvalue")
  limmaout$fdr=p.adjust(limmaout$pvalue,method='fdr',n=length(limmaout$pvalue))
  limmaout$gsea_ranking_score=-log10(limmaout$pvalue)*sign(limmaout$log2fc)
  write.table(limmaout,file=paste("limma_DEG_",cons[i],"_all_genes.txt",sep=""),row.names=FALSE,col.names=TRUE,quote=FALSE,sep="\t")
  
  
  if ((SPECIES=="hg19" | SPECIES=="hg38" | SPECIES=="hg38_30" | SPECIES=="mm10" | SPECIES=="mm10_M21" | SPECIES=="mm9")&&(grepl("genes",params$dtype))) {    
    system(paste0("cd ",getwd(),"; python ../Scripts/FC2karyo.py limma_DEG_",cons[i],"_all_genes.txt 1 3 ",SPECIES," ",KARYOBEDS," > limma_DEG_",cons[i],".input.karyo"))
    system(paste0("cd ",getwd(),"; python ../Scripts/karyoplot.py limma_DEG_",cons[i],".input.karyo ",SPECIES))
    ki1=paste0("limma_DEG_",cons[i],".input.karyo1.png")
    ki2=paste0("limma_DEG_",cons[i],".input.karyo2.png") 
  }
  
  if (SPECIES=="hg19" | SPECIES=="hg38" | SPECIES=="hg38_30" | SPECIES=="mm10" | SPECIES=="mm10_M21" | SPECIES=="mm9" | SPECIES=="canFam3" | SPECIES=="Mmul_8.0.1" ) {
    system(paste0("cd ",getwd(),";/data/CCBR_Pipeliner/db/PipeDB/Conda/envs/r_env/bin/Rscript ../Scripts/karyoploter.R -o limma -d limma_DEG_",contras[i],"-",contras[i+1],"_all_genes.txt -c ", params$karyotext, " -g ", SPECIES))
  }
  
}

```

```{r table, echo=FALSE, warning=FALSE,message=FALSE,results='asis', fig.width=6, fig.height=7}

# DEG  table
colnames(mfc)=paste(cons,"FC",sep="_")
colnames(mpval)=paste(cons,"pval",sep="_")
res=as.data.frame(cbind(mfc,mpval))
rownames(res)=rownames(all.genes.con)
write.table(res,file="limma_res_all_contrasts.txt",sep="\t",col.names=NA)
#DT::datatable(res, caption=htmltools::tags$caption("DEG for all contrasts",style="color:blue; font-size: 18px" ))
DT::datatable(limmaout, extensions = 'Buttons', options = list(dom = 'Bfrtip', buttons = c('colvis','excel', 'csv'), pageLength = 15))
```

## **Volcano Plots** {.tabset}

### **p-value**

```{r p_volcano, echo=FALSE, warning=FALSE,message=FALSE,fig.width=6, fig.height=7}
l
```

### FDR

```{r fdr_volcano, echo=FALSE, warning=FALSE,message=FALSE,fig.width=6, fig.height=7}
v_fdr
```

## **Karyoplots: Significant Genes** {.tabset .tabset-fade .tabset-pills}

### **Page1**

```{r karyo_gene1, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}
if (SPECIES=="hg19" | SPECIES=="hg38" | SPECIES=="hg38_30" | SPECIES=="mm10" | SPECIES=="mm10_M21" | SPECIES=="mm9" | SPECIES=="canFam3" | SPECIES=="Mmul_8.0.1" ){
  cat("![](limma_karyoplot1.png)")
} else {
  cat("Genome Not Supported")
}

```

### Page2

```{r karyo_gene2, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}
if (SPECIES=="hg19" | SPECIES=="hg38" | SPECIES=="hg38_30" | SPECIES=="mm10" | SPECIES=="mm10_M21" | SPECIES=="mm9" | SPECIES=="canFam3" | SPECIES=="Mmul_8.0.1" ){
  cat("![](limma_karyoplot2.png)")
} else {
  cat("Genome Not Supported")
}
```


## **Karyoplots: Binned Expression** {.tabset .tabset-fade .tabset-pills}

### **Page1**

```{r karyo_bin1, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}
if ((SPECIES=="hg19" | SPECIES=="hg38" | SPECIES=="hg38_30" | SPECIES=="mm10" | SPECIES=="mm10_M21" | SPECIES=="mm9")&&(grepl("genes",params$dtype))){
  cat("![](",ki1,")")
} else {
  cat("Genome Not suporrted")
}
```

### Page2

```{r karyo_bin2, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}
if ((SPECIES=="hg19" | SPECIES=="hg38" | SPECIES=="hg38_30" | SPECIES=="mm10" | SPECIES=="mm10_M21" | SPECIES=="mm9")&&(grepl("genes",params$dtype))){
  cat("![](",ki2,")")
} else {
  cat("Genome Not Supported")
}
```

## **Enriched Pathways** {.tabset}

### **Up-Regulated**

```{r l2p_up, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}
if (SPECIES=="hg19" | SPECIES=="mm10" | SPECIES=="mm10_M21" | SPECIES=="hg38" | SPECIES=="hg38_30"){
  pup
} else {
  cat("Genome Not Supported")
}
```

### Down-Regulated

```{r l2p_down, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}
if (SPECIES=="hg19" | SPECIES=="mm10" | SPECIES=="mm10_M21" | SPECIES=="hg38" | SPECIES=="hg38_30"){
  pdw
} else {
  cat("Genome Not Supported")
}
```
