---
title: "limma results"
author: "CCBR RNAseq pipeline"
output: html_document
params:
  folder: result
  sampleinfo: sampletable.txt
  data: RawCountFile_genes_filtered.txt
  contrasts: "GN GP"
  species: hg19
  projectId: "<projectId>"
  projectDesc: "<desc>"
  gtffile: "gencode.v19.annotation.gtf"
  dtype: "<count type>"
  karyobeds: "path/to/karyobedfiles"
---

```{r, include=FALSE, warning=FALSE, message=FALSE}
countMethod <- "subread"
if(grepl("RSEM",params$data)){
  countMethod <- "rsem"
}
projectId<-params$projectId
projectDesc<-params$projectDesc
dateandtime<-format(Sys.time(), "%a %b %d %Y - %X")
```
### **Project:**
####    *`r projectId`*
### **Description:** 
####    *`r projectDesc`*
###  **Count method:** 
####    *`r countMethod`* 
### **Report generated:** 
####    *`r dateandtime`* 

## **Histogram after normalization**


```{r setup, echo=FALSE, warning=FALSE,message=FALSE}

library(rgl)
library(knitr)
library('statmod')
library('RColorBrewer') 
library(gplots)
library('reshape') 
library('ggplot2')
library('limma')
library('geneplotter')
library('ggfortify')
library(amap)
library(DT)
library(plotly)

knitr::opts_chunk$set(fig.width=12, fig.height=8) 
knit_hooks$set(rgl = function(before, options, envir) {
  if (!before) {
    ## after a chunk has been evaluated
    if (rgl.cur() == 0) return()  # no active device
    name = paste(options$fig.path, options$label, sep = '')
    rgl.snapshot(paste(name, '.png', sep = ''), fmt = 'png')
    return(paste('\\includegraphics{', name, '}\n', sep = ''))
  }
})

knit_hooks$set(webgl = hook_webgl)
```


```{r, echo=FALSE, warning=FALSE,message=FALSE}
## library('edgeR')


## grab args

DIR <- params$folder
FILE1 <- params$sampleinfo
FILE2 <- params$data
CONTRASTS <- params$contrasts
SPECIES <-params$species
KARYOBEDS <-params$karyobeds

 
# read files
sampleinfo=read.delim(FILE1)
x = read.delim(FILE2,row.names=1)
colnames(x)=as.character(sampleinfo[,4])
# sampleFiles=as.character(sampleinfo[,2])
Group <- factor(sampleinfo$condition)
design=model.matrix(~0+Group)
contras=unlist(strsplit(CONTRASTS, split=" "))  
# opts_knit$set(root.dir = DIR)
# setwd(DIR)
cat(contras,"\t",length(contras),"\t",contras[1],"\t",contras[2],"\n",file="readcontralimma.txt")

cons=c()
for(i in seq(1, length(contras), by = 2))
{{
cons=c(cons,paste(contras[i],"-",contras[i+1],sep=""))
}}
#print(x)
#print(design)

#png("VoomPlot.png")
v1 <- voom(as.matrix(x),design,plot=TRUE,normalize="quantile")
#dev.off()

sf = v1$E/log2((x/colSums(x))*1000000)

rn=rownames(v1$E)
ensID=apply(array(as.character(rn)),1,function(z) unlist(strsplit(z, "\\|"))[1])
gene=apply(array(as.character(rn)),1,function(z) unlist(strsplit(z, "\\|"))[2])
mydata=cbind(ensID,gene,v1$E)

sf=cbind(ensID,gene,sf)

write.table(sf,file="limma_voom_scaling_factors.txt",row.names=TRUE,col.names=NA,sep="\t",quote=FALSE)
write.table(mydata,file="limma_oom_normalized_data.txt",row.names=TRUE,col.names=NA,sep="\t",quote=FALSE)

#png("HistlimmavoomNormFilter.png")
df.n <- melt(as.data.frame(v1$E))
print(ggplot(df.n) + geom_density(aes(x = value,colour = variable)) + labs(x = NULL) + theme(legend.position='right') + ggtitle("Normalized Counts"))
# dev.off()

```
```{r, echo=FALSE, warning=FALSE,message=FALSE,include=FALSE}
## MDS
png("limma_MDS.png")
## MDS <- plotMDS(v1,xlim=c(-5,5),ylim=c(-5,5),cex=1,pch=20)
plotMDS(v1,xlim=c(-5,5),ylim=c(-5,5),cex=1,pch=20)
## png("limma_MDS.png")
shortname=paste(substr(colnames(v1$E),1,22))
#text(MDS, labels=shortname, cex=0.5, pos=1)
# print(MDS)
## dev.copy(png, paste("limma_MDS.png"))
dev.off()

## 
```

## **Boxplot and PCA**

```{r, set-options, echo=FALSE,webgl=TRUE,message=FALSE}
#options(out.width = 160)

# library(rglwidget)

#colnames(v1$E)=fc1filtnames
df.m <- melt(as.data.frame(v1$E))

#dev.off()
par(mar=c(10,7,1,1))
boxplot(value~variable,las=2,data=df.m,main="Normalized Signal", 
    ylab="Counts",col=as.numeric(as.factor(sampleinfo$condition)))

edf=as.matrix(v1$E)
tedf= t(edf)
tedf=tedf[,apply(tedf,2,var)!= 0]
pca=prcomp(tedf,scale.=T)
tedf1 = data.frame(tedf)
Phenotype=sampleinfo$condition
cell_rep=sampleinfo$label
tedf1$group = as.factor(Phenotype)

pc1 = round(pca$sdev[1]^2/sum(pca$sdev^2)*100,2)
pc2 = round(pca$sdev[2]^2/sum(pca$sdev^2)*100,2)
pc3 = round(pca$sdev[3]^2/sum(pca$sdev^2)*100,2)

pcafactor = as.factor(sampleinfo$condition)

library(RColorBrewer)

col <- brewer.pal(nlevels(pcafactor), "Paired")

p <- plot_ly(as.data.frame(pca$x[,1:3]), x = ~PC1, y = ~PC2, z = ~PC3, color = pcafactor, colors = col, hoverinfo="text",
             hovertext = ~sampleinfo$label) %>%
  add_markers() %>%
  layout(title = "PCA plot", 
         scene = list(xaxis = list(title = paste0("PC1 (",pc1,"%)")),
                      yaxis = list(title = paste0("PC2 (",pc2,"%)")),
                      zaxis = list(title = paste0("PC3 (",pc3,"%)"))))

saveWidget(as_widget(p), "MainPcaPlot.html")

p






# plot(pca,type="lines")  #Decide how many PC's are relevant for plotting
  #pca$x[,1:3]  #look at first 3 PC's

#plot3d(pca$x[,1:3],col = as.integer(tedf1$group),type="s",size=2)
#group.v<-as.vector(cell_rep)
#text3d(pca$x, pca$y, pca$z, group.v, cex=1.0, adj = 1.2) 
# rgl.postscript("pca3d_limma.pdf","pdf")

```

## **Similarity Heatmap**
```{r, echo=FALSE, warning=FALSE,message=FALSE,include=FALSE}

# PCA 2D
pr2=prcomp(tedf,scale.=T)
condition = factor(sampleinfo$condition)
dd=cbind(tedf,condition=as.character(condition))

png("limma_PCA_2D.png")
autoplot(pr2,data=dd, colour = 'condition')
dev.off()
```
```{r, echo=FALSE,warning=FALSE,message=FALSE}
#library(lattice) 

d=Dist(tedf,method="pearson",diag=TRUE)
m=as.matrix(d)

new.palette=colorRampPalette(c("black","red","yellow","white"),space="rgb")
#levelplot(m[1:ncol(m),ncol(m):1],col.regions=new.palette(20))
heatmap(m,symm=TRUE,col=new.palette(20))

```

## **Differentially expressed genes (DEG) for all contrasts: Volcano plots, Pathways and DEG tables**

```{r, echo=FALSE, warning=FALSE,message=FALSE,results='asis'}
library(ggplot2)
library(plotly)
## DEG
nb=length(contras)/2
colnames(design) <- levels(Group)
fit <- lmFit(v1,design)
contrast.matrix <- makeContrasts(contrasts=cons,levels=design)
fitb <- contrasts.fit(fit, contrast.matrix)
ebayes.fit=eBayes(fitb)

#logFC = ebayes.fit$coefficients
#colnames(logFC)=paste(colnames(ebayes.fit$coefficients),"logFC",sep="_")
#pval=ebayes.fit$p.value
#colnames(pval)=paste(colnames(ebayes.fit$p.value),"pval",sep="_")
#res=as.data.frame(cbind(logFC,pval))
#write.table(res,file="limma_res_all_contrasts.txt",sep="\t",col.names=NA)
#DT::datatable(res)

#kp1 = htmltools::tagList()
#kp2 = htmltools::tagList()
l <- htmltools::tagList()
pup = htmltools::tagList()
pdw = htmltools::tagList()
mfc=c()
mpval=c()

#if ((grepl("hg",SPECIES)||grepl("mm",SPECIES))&&(grepl("genes",params$dtype)))
#{
#system(paste0("cd ",getwd(),"; perl ../Scripts/gtf2bed.pl ",params$gtffile," |sort -k1,1 -k2,2n > karyobed.bed"))
#system(paste0("cd ",getwd(),"; cut -f1 karyobed.bed|uniq > chrs.txt; while read a ;do cat karyobed.bed | awk -F \"\\t\" -v a=$a \'{if ($1==a) {print}}\' > karyobed.${a}.bed;done < chrs.txt"))
#}

for (i in 1:nb)
{{
all.genes.con = topTable(ebayes.fit, coef = i, number=nrow(ebayes.fit), sort.by="none")
## new staff ------

all.genes.con$FC <- ifelse(all.genes.con$logFC<0, -1/(2^all.genes.con$logFC), 2^all.genes.con$logFC)
final=all.genes.con
mfc=cbind(mfc,final$FC)
mpval=cbind(mpval,final$P.Value)

## pathway enrichment --------------------------------------------
if (SPECIES=="hg19" | SPECIES=="mm10" | SPECIES=="hg38")
{
#gup=gene[which(final$PValue<0.05 & final$FC >=1.5)]
#gdw=gene[which(final$PValue<0.05 & final$FC <=-1.5)]
iup=which(final$P.Value<0.05 & final$FC >=0)
idw=which(final$P.Value<0.05 & final$FC <0)
fin.up=final[iup,]
if (length(iup) > 50)
  {
   fin.up=fin.up[order(fin.up$P.Value),]
   fin.up=fin.up[1:50,]
  }
x2=rownames(fin.up)
gup=apply(array(as.character(x2)),1,function(z) unlist(strsplit(z, "\\|"))[2])

fin.dw=final[idw,]
if (length(idw) > 50)
  {
   fin.dw=fin.dw[order(fin.dw$P.Value),]
   fin.dw=fin.dw[1:50,]
  }
x2=rownames(fin.dw)
gdw=apply(array(as.character(x2)),1,function(z) unlist(strsplit(z, "\\|"))[2])

extension=paste0("_",cons[i],".txt")
if (SPECIES=="hg19" | SPECIES=="hg38")
{
write.table(gup,paste0("limma_gene_up_path",extension),quote=F,row.names=F,col.names=F)
write.table(gdw,paste0("limma_gene_dw_path",extension),quote=F,row.names=F,col.names=F)
}
else
{
write.table(gup,"limma_temp_up.txt",quote=F,row.names=F,col.names=F)
write.table(gdw,"limma_temp_dw.txt",quote=F,row.names=F,col.names=F)
system(paste0("cat limma_temp_up.txt | grep -v \"^NA\" | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/m2h | grep -v XXXX | cut -f2 -d\" \"  >limma_gene_up_path",extension))
system(paste0("cat limma_temp_dw.txt | grep -v \"^NA\" | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/m2h | grep -v XXXX | cut -f2 -d\" \"  >limma_gene_dw_path",extension))
}
# extension=paste0("_",contras[i],"_vs_",contras[i+1],".txt")
system(paste0("cat limma_gene_up_path",extension," |sort | uniq | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/l2p >limma_res_path_up",extension))
system(paste0("cat limma_gene_dw_path",extension," |sort | uniq | /data/CCBR_Pipeliner/db/PipeDB/bin/L2P/l2p >limma_res_path_dw",extension))

mup=read.delim(paste0("limma_res_path_up",extension),header=F)
mdw=read.delim(paste0("limma_res_path_dw",extension),header=F)
colnames(mup)=c("pval","fdr","ratio","nb.hits","nb.genes.path","nb.user.genes","tot.back.genes","path_id","source","description","type","gene.list")
colnames(mdw)=c("pval","fdr","ratio","nb.hits","nb.genes.path","nb.user.genes","tot.back.genes","path_id","source","description","type","gene.list")

#mup=mup[which(mup[,1]<0.05),]
#mdw=mdw[which(mdw[,1]<0.05),]

mup=mup[which(mup[,1]<0.05 & mup[,4]>=5),]
mdw=mdw[which(mdw[,1]<0.05 & mdw[,4]>=5),]

mup=mup[order(mup[,1]),]
mdw=mdw[order(mdw[,1]),]
mup=mup[,c(8,9,10,11,1:7,12)]
mdw=mdw[,c(8,9,10,11,1:7,12)]


pup[[i]] <- DT::datatable(mup, caption=htmltools::tags$caption(paste0("Pathways for top 50 up-regulated genes ",cons[i]) ,style="color:red; font-size: 18px" ),rownames=F)
pdw[[i]] <- DT::datatable(mdw, caption=htmltools::tags$caption(paste0("Pathways for top 50 down-regulated genes ",cons[i]) ,style="color:green; font-size: 18px" ),rownames=F )

}
## ------------------------------------------------------------------



log_FC=all.genes.con$logFC
log_pval=-log10(all.genes.con$P.Value)
Significant=rep("NotSignificant",length(log_FC))
Significant[which(all.genes.con$P.Value<0.05 & abs(all.genes.con$logFC)>=1)]="Significant&LogFoldChange"
Significant[which(all.genes.con$P.Value<0.05 & abs(all.genes.con$logFC)<1)]="Significant"
Significant[which(all.genes.con$P.Value>=0.05 & abs(all.genes.con$logFC)>=1)]="LogFoldChange"
gene=rownames(all.genes.con)
volcano_data=as.data.frame(cbind(gene,log_FC,log_pval,Significant))
l[[i]] <- plot_ly(data = volcano_data, x = log_FC, y = log_pval, text = gene, mode = "markers", color = Significant) %>% layout(title =paste("Volcano Plot for ",cons[i],sep=""),xaxis=list(title="Fold Change",range =c(-5,5),tickvals=c(-5,-4,-3,-2,-1,0,1,2,3,4,5),ticktext=c('-32','-16','-8','-4','-2','1','2','4','8','16','32')),yaxis=list(title="-Log10 pvalue",range =c(0,10)))

#l[[i]] <- plot_ly(data = volcano_data, x = log_FC, y = log_pval, text = gene, mode = "markers", color = Significant) %>% layout(title =paste("Volcano Plot for ",cons[i],sep=""),xaxis=list(title="Log2_FC",range =c(-10,10)),yaxis=list(title="-Log10_pvalue",range =c(0,10)))
#  %>% layout(annotations = a)

# DT::datatable(volcano_data)
#MAplot <- plot(ebayes.fit,coef=i)
#print(MAplot)
#dev.copy(png, paste(cons[i],"_MAplot_limma_old.png",sep=""))
#dev.off()
dataf=data.frame("m"=all.genes.con$AveExpr,"fc"=all.genes.con$logFC,"sig"=all.genes.con$adj.P.Val<0.05)
png(paste(cons[i],"_MAplot_limma_v2.png",sep=""))
plotMA(dataf,log="",main=cons[i],ylim=range(all.genes.con$logFC))
dev.off()
# all.genes.con$FC <- ifelse(all.genes.con$logFC<0, -1/(2^all.genes.con$logFC), 2^all.genes.con$logFC)
x=rownames(all.genes.con)
ensID=apply(array(as.character(x)),1,function(z) unlist(strsplit(z, "\\|"))[1])
gene=apply(array(as.character(x)),1,function(z) unlist(strsplit(z, "\\|"))[2])
all.genes.con=cbind(ensID,gene,all.genes.con)
write.table(all.genes.con,file=paste("limma_DEG_",cons[i],"_all_genes.txt",sep=""),sep="\t",col.names=NA)

if ((grepl("hg",SPECIES)||grepl("mm",SPECIES))&&(grepl("genes",params$dtype)))
{
system(paste0("cd ",getwd(),"; python ../Scripts/FC2karyo.py limma_DEG_",cons[i],"_all_genes.txt 2 3 ",SPECIES," ",KARYOBEDS," > limma_DEG_",cons[i],".input.karyo"))
system(paste0("cd ",getwd(),"; python ../Scripts/karyoplot.py limma_DEG_",cons[i],".input.karyo ",SPECIES))
ki1=paste0("limma_DEG_",cons[i],".input.karyo1.png")
ki2=paste0("limma_DEG_",cons[i],".input.karyo2.png")
cat("\n###KARYO PLOT",cons[i],"\n")
cat("![](",ki1,")")
#kp1[[i]]<-knitr::include_graphics(paste0("limma_DEG_",cons[i],".input.karyo1.png"))
cat("\n")
cat("![](",ki2,")")
#kp2[[i]]<-knitr::include_graphics(paste0("limma_DEG_",cons[i],".input.karyo2.png"))
cat("\n\n")
}


}}

# add DEG  table
colnames(mfc)=paste(cons,"FC",sep="_")
colnames(mpval)=paste(cons,"pval",sep="_")
res=as.data.frame(cbind(mfc,mpval))
rownames(res)=rownames(all.genes.con)
write.table(res,file="limma_res_all_contrasts.txt",sep="\t",col.names=NA)

#kp1[[i]]
#kp2[[i]]

l

pup

pdw

DT::datatable(res, caption=htmltools::tags$caption("DEG for all contrasts",style="color:blue; font-size: 18px" ))

```


```{r, echo=TRUE, message=FALSE, warning=FALSE}

sessionInfo()

```

