---
title: "Inital/QC Report"
author: "CCBR scRNAseq pipeline"
output: html_document
params:
  matrix: "path/to/count/matrix"
  mattype: "cellranger"
  species: "hg19"
  projectId: "<projectId>"
  projectDesc: "<desc>"
  doCycleRegress: "TRUE"
---

```{r headers, include=FALSE, warning=FALSE, message=FALSE}
projectId<-params$projectId
projectDesc<-params$projectDesc
dateandtime<-format(Sys.time(), "%a %b %d %Y - %X")
species<-params$species
matrix<-params$matrix
mattype <- params$mattype
doCycleRegress <- params$doCycleRegress
```

### **Project:**
####    *`r projectId`*
### **Description:** 
####    *`r projectDesc`*
### **Report generated:** 
####    *`r dateandtime`* 

```{r setup, echo=FALSE, warning=FALSE,message=FALSE,fig.keep='all'}
# library(rgl)
library(knitr) #
library(Seurat) #
library(dplyr)
library(Matrix) #
library(scater) #
library(DropletUtils) #
library(scran) #
library(limma)
library(GSVA) #
library(GSEABase) #
library(genefilter) #
library(AnnotationDbi)
library(org.Mm.eg.db)
library(org.Hs.eg.db) #
library(gplots) #
library(gdata) #
library(RColorBrewer)
library(scales)
library(reshape2)
library(ggplot2)

# knit_hooks$set(rgl = function(before, options, envir) {
#   if (!before) {
#     ## after a chunk has been evaluated
#     if (rgl.cur() == 0) return()  # no active device
#     name = paste(options$fig.path, options$label, sep = '')
#     rgl.snapshot(paste(name, '.png', sep = ''), fmt = 'png')
#     return(paste('\\includegraphics{', name, '}\n', sep = ''))
#   }
# })

#knit_hooks$set(webgl = hook_webgl)

#READS IN COUNTS MATRIX AS SINGLE CELL EXPRESSION SET OBJECT
#THIS IS NEEDED FOR ANALYSIS IN SCATER
calling.cells=FALSE
if(mattype=="cellranger"){
  sce <- read10xCounts(matrix)
}else if(mattype=="cellranger_raw"){
  calling.cells=TRUE
  sce <- read10xCounts(matrix)
  is.mito <- grepl("^MT-",rowData(sce)@listData$Symbol)|
    grepl("^mt-",rowData(sce)@listData$Symbol)|
    grepl("^Mt-",rowData(sce)@listData$Symbol)
  sce <- calculateQCMetrics(sce, feature_controls=list(Mt=is.mito))
  bcrank <- barcodeRanks(counts(sce))
  #THE FOLLOWING GENERATES A BARCODE RANK PLOT
  #highlights percent mitochondrial content
  uniq <- !duplicated(bcrank$rank)
  bcplot <- data.frame(rank=bcrank$rank[uniq], total=bcrank$total[uniq])
  bcplot_mito <- ggplot(bcplot, aes(rank,total)) + 
    scale_x_continuous(trans='log10') +
    scale_y_continuous(trans='log10') +
    geom_point(shape=1, aes(colour=sce$pct_counts_Mt[uniq])) +
    scale_colour_gradient2(limits = c(0,10), high="red") +
    xlab("Rank") +
    ylab("Total UMI count") +
    geom_hline(yintercept=bcrank$inflection, linetype="dashed", col="darkgreen") +
    geom_text(aes(0,bcrank$inflection,label="Inflection"),vjust=-1, hjust=0) +
    geom_hline(yintercept=bcrank$knee, linetype="dashed", col="dodgerblue") +
    geom_text(aes(0,bcrank$knee,label="Knee"),vjust=-1, hjust=0) +
    ggtitle(label = "Barcode Rank Plot (w/ pct mito)") +
    theme(legend.title = element_blank())
  #THE FOLLOWING GENERATES A BARCODE RANK PLOT USING CELL CALLING
  #The cell calling algorithm in Droplet Utils as described in Lun et al.
  set.seed(100)
  e.out <- emptyDrops(counts(sce))
  sce$calledcell <- FALSE
  sce$calledcell[which(e.out$FDR <= 0.01)] <- TRUE
  bcplot_caller <- ggplot(bcplot, aes(rank,total)) + 
    scale_x_continuous(trans='log10') +
    scale_y_continuous(trans='log10') +
    geom_point(shape=1, aes(colour=sce$calledcell[uniq])) +
    xlab("Rank") +
    ylab("Total UMI count") +
    geom_hline(yintercept=bcrank$inflection, linetype="dashed", col="darkgreen") +
    geom_text(aes(0,bcrank$inflection,label="Inflection"),vjust=-1, hjust=0) +
    geom_hline(yintercept=bcrank$knee, linetype="dashed", col="dodgerblue") +
    geom_text(aes(0,bcrank$knee,label="Knee"),vjust=-1, hjust=0) +
    ggtitle(label = "Barcode Rank Plot (cell calling)") +
    theme(legend.title = element_blank())
  
  print(bcplot_mito)
  print(bcplot_caller)
  
  sce <- sce[,which(sce$calledcell)]
  calledcells <- sce@colData$Barcode
  if (all(grepl(pattern = "\\-1$", x = calledcells))) {
	  calledcells <- 
	    as.vector(x = as.character(x = sapply(X = calledcells, FUN = ExtractField, field = 1, delim = "-")))
  }
}else if(mattype=="zumi"){
  calling.cells=TRUE
  zumi.output <- readRDS(matrix)
  umicounts <- zumi.output$exons$umicounts
  sce <- SingleCellExperiment(list(counts=umicounts))
  isSpike(sce, "MySpike") <- grep("^ERCC", rownames(sce))
  sce@colData$Barcode <- colnames(sce)
  rowData(sce)$Symbol <- rownames(sce)
  
  is.mito <- grepl("^MT-",rowData(sce)@listData$Symbol)|
    grepl("^mt-",rowData(sce)@listData$Symbol)|
    grepl("^Mt-",rowData(sce)@listData$Symbol)
  sce <- calculateQCMetrics(sce, feature_controls=list(Mt=is.mito))
  bcrank <- barcodeRanks(counts(sce))
  #THE FOLLOWING GENERATES A BARCODE RANK PLOT
  #highlights percent mitochondrial content
  uniq <- !duplicated(bcrank$rank)
  bcplot <- data.frame(rank=bcrank$rank[uniq], total=bcrank$total[uniq])
  bcplot_mito <- ggplot(bcplot, aes(rank,total)) + 
    scale_x_continuous(trans='log10') +
    scale_y_continuous(trans='log10') +
    geom_point(shape=1, aes(colour=sce$pct_counts_Mt[uniq])) +
    scale_colour_gradient2(limits = c(0,10), high="red") +
    xlab("Rank") +
    ylab("Total UMI count") +
    geom_hline(yintercept=bcrank$inflection, linetype="dashed", col="darkgreen") +
    geom_text(aes(0,bcrank$inflection,label="Inflection"),vjust=-1, hjust=0) +
    geom_hline(yintercept=bcrank$knee, linetype="dashed", col="dodgerblue") +
    geom_text(aes(0,bcrank$knee,label="Knee"),vjust=-1, hjust=0) +
    ggtitle(label = "Barcode Rank Plot (w/ pct mito)") +
    theme(legend.title = element_blank())
  #THE FOLLOWING GENERATES A BARCODE RANK PLOT USING CELL CALLING
  #The cell calling algorithm in Droplet Utils as described in Lun et al.
  set.seed(100)
  e.out <- emptyDrops(counts(sce))
  sce$calledcell <- FALSE
  sce$calledcell[which(e.out$FDR <= 0.01)] <- TRUE
  bcplot_caller <- ggplot(bcplot, aes(rank,total)) + 
    scale_x_continuous(trans='log10') +
    scale_y_continuous(trans='log10') +
    geom_point(shape=1, aes(colour=sce$calledcell[uniq])) +
    xlab("Rank") +
    ylab("Total UMI count") +
    geom_hline(yintercept=bcrank$inflection, linetype="dashed", col="darkgreen") +
    geom_text(aes(0,bcrank$inflection,label="Inflection"),vjust=-1, hjust=0) +
    geom_hline(yintercept=bcrank$knee, linetype="dashed", col="dodgerblue") +
    geom_text(aes(0,bcrank$knee,label="Knee"),vjust=-1, hjust=0) +
    ggtitle(label = "Barcode Rank Plot (cell calling)") +
    theme(legend.title = element_blank())
  
  print(bcplot_mito)
  print(bcplot_caller)
  
  sce <- sce[,which(sce$calledcell)]
  calledcells <- sce@colData$Barcode

}else if(mattype=="biorad"){ #THIS IS FOR BIORAD/DENSE COUNT TABLE INPUT
  counts.table <- read.delim(matrix, stringsAsFactors = F, header = TRUE, sep=",", row.names=1)
  counts.matrix <- t(as.matrix(counts.table))
  sce <- SingleCellExperiment(list(counts=counts.matrix))
  isSpike(sce, "MySpike") <- grep("^ERCC", rownames(sce))
  sce@colData$Barcode <- colnames(sce)
  rowData(sce)$Symbol <- rownames(sce)
}else{
  stop("Input type not recognized!")
}

#The following calculates various metrics for gene and cell filtering
#Genes that have expression in at least 0.1 percent of total # of cells
#And cells are filtered on low ngenes, low nUMI, and high %mito
#Cell filtering is done based off of distribution of data (median +/- 3 deviations)
numcells <- nexprs(sce, byrow=TRUE)
keepgenes <- numcells >= ifelse(round(0.001*ncol(sce))!=0,round(0.001*ncol(sce)),1)
sce <- sce[keepgenes,]
numdroppedgenes <- sum(!keepgenes)
totalgenes <- length(keepgenes)
is.mito <- grepl("^MT-",rowData(sce)@listData$Symbol)|grepl("^mt-",rowData(sce)@listData$Symbol)|grepl("^Mt-",rowData(sce)@listData$Symbol)
sce <- calculateQCMetrics(sce, feature_controls=list(Mt=is.mito))
libsize.drop <- isOutlier(sce$total_counts, nmads=3, type="lower", log=TRUE)
feature.drop <- isOutlier(sce$total_features_by_counts, nmads=3, type="lower", log=TRUE)
mito.drop <- isOutlier(sce$pct_counts_Mt, nmads=3, type="higher")
prefilterstats <- data.frame(row.names=sce@colData$Barcode, counts=sce$total_counts, features=sce$total_features_by_counts, pct.mito=sce$pct_counts_Mt)
write.table(prefilterstats,paste0(projectId,"_prefilter_metrics.txt"),sep="\t",quote=F,col.names=NA)
sce$keep <- !(libsize.drop | feature.drop | mito.drop)
#STORE THE FILTERED DATASET IN A DIFFERENT OBJECT FOR NOW
sce_filtered <- sce[,!(libsize.drop | feature.drop | mito.drop)]
keepcells <- sce_filtered@colData$Barcode
droppedcells <- setdiff(sce@colData$Barcode,keepcells)
#Cellranger automatically formats the barcodes using the following lines of code
#Have to execute the same lines on the vector of barcodes to keep in 'keepcells'
#before passing it along to cellranger
if (all(grepl(pattern = "\\-1$", x = keepcells))) {
	keepcells <- as.vector(x = as.character(x = sapply(X = keepcells, FUN = ExtractField, field = 1, delim = "-")))
}

#Import data into Seurat using the following
#The min.cells parameter is used here for the gene filtering
if(mattype=="cellranger"){
  so.data <- Read10X(matrix)
  so <- CreateSeuratObject(so.data, 
                         min.genes = 0, 
                         min.cells=ifelse(round(0.001*ncol(sce))!=0,round(0.001*ncol(sce)),1), 
                         project=projectId)
}else if(mattype=="cellranger_raw"){
  so.data <- Read10X(matrix)
  #WITH RAW CELLRANGER DATA, have to FIRST EXTRACT ONLY CALLED CELLS
  #As determined by the Lun et al. method from the DropletUtils step
  so.data <- so.data[,calledcells]
  so <- CreateSeuratObject(so.data, 
                         min.genes = 0, 
                         min.cells=ifelse(round(0.001*ncol(sce))!=0,round(0.001*ncol(sce)),1), 
                         project=projectId)
}else if(mattype=="zumi"){
  so <- CreateSeuratObject(raw.data = umicounts, min.genes = 0,
                           min.cells=ifelse(round(0.001*ncol(sce))!=0,round(0.001*ncol(sce)),1),
                           project=projectId)
}else if(mattype=="biorad"){
  so <- CreateSeuratObject(raw.data = counts.matrix, 
                           min.genes = 0, 
                           min.cells=ifelse(round(0.001*ncol(sce))!=0,round(0.001*ncol(sce)),1), 
                           project=projectId)
  so <- MakeSparse(so)
}
# so@raw.data <- so@data
# so <- Setup(so, min.genes=0, min.cells=round(0.001*ncol(sce)), do.logNormalize=T, do.scale=F, do.center=F, total.expr=1e4, project=projectId, save.raw=F)
```
`r if(calling.cells){paste0(" \nThere are **",length(calledcells),"** called cells in this sample.\n")}`

### **nGenes vs nUMI** (Pre-Filter)

```{r nUMI_plot, echo=FALSE, message=FALSE, warning=FALSE}
#THIS NGENE VS NUMI PLOT IS USEFUL FOR FINDING DROPLETS WITH MULT. CELLS
#xthresh = median(so@data.info$nUMI)+3*stats::mad(so@data.info$nUMI)
#ythresh = median(so@data.info$nGene)+3*stats::mad(so@data.info$nGene)
GenePlot(so, "nUMI", "nGene")
so <- SubsetData(so, cells.use=keepcells)
so <- NormalizeData(so, normalization.method="LogNormalize", scale.factor = 10000)
```

### **Pre-Filter Histograms**

```{r histograms_1, echo=FALSE,warning=FALSE,message=FALSE}
#The following three chunks generates the prefilter histograms
par(mfrow=c(1,2))
hist(sce$total_counts, xlab="Library size", main="", 
    breaks=20, col="grey80", ylab="Number of cells")
plot(density(sce$total_counts),col="red", xlab="Library size", main="")
```
```{r histograms_2, echo=FALSE,warning=FALSE,message=FALSE}
par(mfrow=c(1,2))
hist(sce$total_features_by_counts, xlab="Number of expressed genes", main="", 
    breaks=20, col="grey80", ylab="Number of cells")
plot(density(sce$total_features_by_counts),col="red", xlab="Number of expressed genes", main="")
```
```{r histograms_3, echo=FALSE,warning=FALSE,message=FALSE}
par(mfrow=c(1,2))
hist(sce$pct_counts_Mt, xlab="Mitochondrial proportion (%)", 
    ylab="Number of cells", breaks=20, main="", col="grey80")
plot(density(sce$pct_counts_Mt),col="red", xlab="Mitochondrial proportion (%)", main="")
```

### **Removed** `r numdroppedgenes` **genes with expression in less than 0.1% of all cells from** `r totalgenes` **inital total genes. Downstream analysis uses the remaining** `r totalgenes-numdroppedgenes` **genes.**

### **Removed the following** `r length(droppedcells)` **barcoded cells as poor quality. Downstream analysis uses the remaining** `r length(keepcells)` **cells.**


```{r filter_metrics, echo=FALSE, message=FALSE, warning=FALSE, results='asis'}
#OUTPUTS THE NUMBER OF CELLS IDENTIFIED FOR REMOVAL ON EACH FILTERING THRESHOLD
kable(data.frame(ByLibSize=sum(libsize.drop), ByFeature=sum(feature.drop), ByMito=sum(mito.drop), Remaining=ncol(sce_filtered)))
cat("  \n")
#cat(unlist(droppedcells))
#cat("  \n")
```

### **PCA highlighting poor quality cells that are dropped for downstream analysis**

```{r qc_pca, echo=FALSE,warning=FALSE,message=FALSE}
#This PCA plot is generated in scater
#This is just a useful tool to visualize the cells marked as dropped
#The size of each point is also plotted by nGene
#Please note: this uses scater to do PCA, rather than Seurat (will look different)
#Only used for internal evaluation of filtering steps.
sce <- scater::normalize(sce)
sce <- scater::runPCA(sce)
scater::plotPCA(sce,
                size_by = "total_features_by_counts", 
                colour_by = "keep")
sce_filtered <- scater::normalize(sce_filtered)
sce_filtered <- runPCA(sce_filtered)
#GRAB THE FILTERED DATASET FOR POST-FILTER HISTOGRAM GENERATION
sce<-sce_filtered
```

### **Post-Filter Histograms**

```{r histograms_4, echo=FALSE,warning=FALSE,message=FALSE}
par(mfrow=c(1,2))
hist(sce$total_counts, xlab="Library size", main="", 
    breaks=20, col="grey80", ylab="Number of cells")
plot(density(sce$total_counts),col="red", xlab="Library size", main="")
```
```{r histograms_5, echo=FALSE,warning=FALSE,message=FALSE}
par(mfrow=c(1,2))
hist(sce$total_features_by_counts, xlab="Number of expressed genes", main="", 
    breaks=20, col="grey80", ylab="Number of cells")
plot(density(sce$total_features_by_counts),col="red", xlab="Number of expressed genes", main="")
```
```{r histograms_6, echo=FALSE,warning=FALSE,message=FALSE}
par(mfrow=c(1,2))
hist(sce$pct_counts_Mt, xlab="Mitochondrial proportion (%)", 
    ylab="Number of cells", breaks=20, main="", col="grey80")
plot(density(sce$pct_counts_Mt),col="red", xlab="Mitochondrial proportion (%)", main="")
```

### **Cell Cycle Phase**

```{r cell_cycle, echo=FALSE,warning=FALSE,message=FALSE}
#Cell cycle phase assignment is done only if appropriate paramter is set
#Uses Cyclone to do assignment
isHuman = grepl("hg",tolower(species))|grepl("grch",tolower(species))
isMouse = grepl("mm",tolower(species))|grepl("grcm",tolower(species))
if(doCycleRegress){
  if(isHuman|isMouse){
    if(isHuman){
      pairs <- readRDS(system.file("exdata", "human_cycle_markers.rds", package="scran"))
    }
    if(isMouse){
      pairs <- readRDS(system.file("exdata", "mouse_cycle_markers.rds", package="scran"))
    }
    sce_filtered_temp <- sce_filtered
    if(mattype=="biorad"){
      #BIORAD/COUNTS MATRICIES TYPICALLY USE SYMBOLS
      #This step is to convert to ENSEMBL IDs for cyclone
      if(isHuman){
        annotation <- org.Hs.eg.db
      }
      if(isMouse){
        annotation <- org.Mm.eg.db
      }
      genes <- rownames(sce_filtered_temp)
      g2e <- AnnotationDbi::select(x=annotation,
                                   keys=genes,column="ENSEMBL",keytype="SYMBOL")
      #Hack to deal with many-to-one situations in annotation mapping
      g2e <- g2e[!duplicated(g2e$ENSEMBL),]
      g2e$ENSEMBL[is.na(g2e$ENSEMBL)] <- g2e$SYMBOL[is.na(g2e$ENSEMBL)]
      rownames(sce_filtered_temp)<-g2e$ENSEMBL
    }
    if(mattype=="zumi"){
      #remove the version number from ENSEMBL ID in the case of Zumi data
      rownames(sce_filtered_temp) <- gsub("\\..*","",rownames(sce_filtered_temp))
    }
    assignments <- cyclone(sce_filtered_temp, pairs, gene.names=rownames(sce_filtered_temp))
    plot(assignments$score$G1, assignments$score$G2M, xlab="G1 score", ylab="G2/M score", pch=16, 
      col = ifelse((assignments$score$G1>=0.5)&(assignments$score$G2M<0.5),'green','orange'))
    abline(h=0.5,col="red")
    abline(v=0.5,col="red")
    legend("topleft", paste0("[G2/M] ",
           length(which(assignments$score$G2M >= 0.5 & assignments$score$G1 < 0.5))),
           bty="n", text.col="red")
    legend("topright", paste0("[Unknown] ",
           length(which(assignments$score$G1 >= 0.5 & assignments$score$G2M >= 0.5))),
           bty="n", text.col="red")
    legend("bottomleft", paste0("[S] ",
           length(which(assignments$score$G1 < 0.5 & assignments$score$G2M < 0.5))),
           bty="n", text.col="red")
    legend("bottomright", paste0("[G1] ",
           length(which(assignments$score$G1 >= 0.5 & assignments$score$G2M < 0.5))),
           bty="n", text.col="red")
  }
  sce$G1score <- assignments$score$G1
  sce$G2Mscore <- assignments$score$G2M
}
```

### **PCA 1:2 Before Regressing Cell Cycle Phase**

```{r regress_confounding, echo=FALSE,warning=FALSE,message=FALSE,message=FALSE,results='hide',fig.keep='all'}
#THE FOLLOWING ATTACHES METADATA INFORMATION FOR nUMI, %mito, (and cycle stage)
if(doCycleRegress){
  stage <- c()
  for(i in 1:length(sce$G1score)){
    if(isNA(sce$G1score[i]) | isNA(sce$G2Mscore[i]) | (sce$G1score[i] >= 0.5 & sce$G2Mscore[i] >= 0.5)){
      stage[i] <- "Unknown"
      if(isNA(sce$G1score[i])){
        sce$G1score[i] <- 0.5
      }
      if(isNA(sce$G2Mscore[i])){
        sce$G2Mscore[i] <- 0.5
      }
    }else if(sce$G1score[i] >= 0.5){
      stage[i] <- "G1"
    }else if(sce$G2Mscore[i] >= 0.5){
      stage[i] <- "G2M"
    }else{
      stage[i] <- "S"
    }
  }
  metadata <- data.frame(percent.mito=sce$pct_counts_Mt,G1.score=sce$G1score,G2M.score=sce$G2Mscore,stage=stage,row.names=keepcells)
}else{
  metadata <- data.frame(percent.mito=sce$pct_counts_Mt,row.names=keepcells)
}
so <- AddMetaData(so, metadata)
so <- FindVariableGenes(so, mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5, do.plot=F)
so_temp <- ScaleData(object = so, vars.to.regress = c("nUMI", "percent.mito"))
#Note: can set ScaleData to only regress and not scale if this is an issue
# - via setting do.scale=F
so_temp <- RunPCA(so_temp, pc.genes = so_temp@var.genes, do.print = FALSE)
if(doCycleRegress){
  PCAPlot(so_temp, 1, 2, group.by="stage",cols.use=c(rgb(1, 0, 0, 0.7),rgb(0, 1, 0, 0.3),rgb(0, 0, 1, 0.7),rgb(0, 0, 0, 0.7)))
}else{
  PCAPlot(so_temp, 1, 2)
}
```

### **PCA 1:2 After Regressing by G1/G2M Scores**

```{r regress_confounding_final, echo=FALSE,warning=FALSE,message=FALSE,message=FALSE,results='hide',fig.keep='all'}
if(doCycleRegress){
  so <- ScaleData(object = so, vars.to.regress = c("nUMI", "percent.mito", "G1.score", "G2M.score"))
  so <- RunPCA(so, pc.genes = so@var.genes, do.print = FALSE)
  PCAPlot(so, 1, 2, group.by="stage",cols.use=c(rgb(1, 0, 0, 0.7),rgb(0, 1, 0, 0.3),rgb(0, 0, 1, 0.7),rgb(0, 0, 0, 0.7)))
}else{
  so <- so_temp
}
```

### **Highest Expression**

```{r highestexpression, echo=FALSE,warning=FALSE,message=FALSE}
fontsize <- theme(axis.text=element_text(size=12), axis.title=element_text(size=16))
make.names(rowData(sce)$Symbol,unique=TRUE)->rownames(sce)
plotQC(sce, type = "highest-expression", n=20) + fontsize
rowData(sce)$ID->rownames(sce)
```

### **Seurat MeanVar Plot**

```{r Seurat_meanvar, echo=FALSE,warning=FALSE,message=FALSE,results='hide',fig.keep='all'}
so <- FindVariableGenes(so, mean.function = ExpMean, dispersion.function = LogVMR, x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)
```

### **PCA 1:2 QC**
```{r Seurat_pcqc, echo=FALSE,warning=FALSE,message=FALSE,results='hide',fig.keep='all'}
#The following is an implementation of Maggie's functions
#FOR PCA PSEUDOCOLORING

#Select color palette:
#Lab.palette <- colorRampPalette(brewer.pal(8,"Set1"))
#Lab.palette <- colorRampPalette(brewer.pal(8,"Set3"))
Lab.palette <- colorRampPalette(brewer.pal(8,"Set2"))
#Lab.palette <- colorRampPalette(brewer.pal(8,"Accent"))
#Lab.palette <- colorRampPalette(c(rgb(50,50,50,1), rgb(,0,0,0)), alpha = TRUE)
#pie(rep(1,12),col=Lab.palette(12))

draw.PCA.clus <- function(obj,colnum,m){
  objname=obj@project.name
  clusmat=data.frame(pc1=obj@dr$pca@cell.embeddings[,1],pc2=obj@dr$pca@cell.embeddings[,2],
            clusid=as.character(obj@meta.data[[m]]))
  title=as.character(m)
  cols=list()
  n=length(unique((obj@meta.data[[m]])))
  cols[[1]]=brewer.pal(8, "Set3")[-2]  #Alternative
  cols[[2]]=brewer.pal(8, "Set1")[-6]
  cols[[3]]=c(brewer.pal(8,"Set2"),cols[[2]][-2])
  cols[[4]]=c("#F8766D","#FF9912","#a100ff","#00BA38","#619CFF","#FF1493","#010407")
  cols[[5]]=c("blue","red","grey")
  cols[[6]]=Lab.palette(n)
  ggplot(clusmat, aes(x=pc1, y=pc2)) +
    theme_bw() +
    theme(legend.title=element_blank()) +
    geom_point(aes(colour=clusid),size=0.5) +
    scale_color_manual(values=cols[[colnum]]) +
    xlab("pc1") + ylab("pc2") +
    ggtitle(title)
  # z=ifelse (any(SO@calc.params$ScaleData$vars.to.regress %in% 
  #                 c("S.Score","G2M.Score","Prolif.Score")),
  #           paste0("graphics/",objname,"/",m,".pca",".png"),
  #           paste0("graphics/",objname,"/",m,".pca.nr",".png"))
  # ggsave(z, width = 8, height = 5)
}

draw.PCA <- function(obj,colnum,m){
  objname=obj@project.name
  clusmat=data.frame(pc1=obj@dr$pca@cell.embeddings[,1],
                     pc2=obj@dr$pca@cell.embeddings[,2],
                     clusid=obj@meta.data[[m]])
  title=as.character(m)
  cols=list()
  mid=median(obj@meta.data[[m]])
  ggplot(clusmat, aes(x=pc1, y=pc2)) +
    theme_bw() +
    theme(legend.title=element_blank()) +
    geom_point(aes(colour=clusid),size=0.5) +
    scale_color_gradient2(low = muted("blue4"), mid = "white", high = "red",
                          midpoint = mid) + 
    xlab("pc1") + ylab("pc2") +
    ggtitle(title)
  # z=ifelse (any(SO@calc.params$ScaleData$vars.to.regress %in% 
  #                 c("S.Score","G2M.Score","Prolif.Score")),
  #           paste0("graphics/",objname,"/",m,".pca",".png"),
  #           paste0("graphics/",objname,"/",m,".pca.nr",".png"))
  # ggsave(z, width = 8, height = 5)
}

draw.PCA.clus(so,colnum=3,m = "orig.ident")
if(doCycleRegress){
  draw.PCA.clus(so,colnum=3,m = "stage")
  draw.PCA(so,colnum=3,m = "G1.score")
  draw.PCA(so,colnum=3,m = "G2M.score")
}
draw.PCA(so,colnum=3,m = "percent.mito")
draw.PCA(so,colnum=3,m = "nGene")
draw.PCA(so,colnum=3,m = "nUMI")

```


### **Seurat PCA Genes**

```{r Seurat_vizpca, echo=FALSE,warning=FALSE,message=FALSE,results='hide',fig.keep='all',fig.height=70}
VizPCA(so, 1:20, nCol=2) #change PCs from 1:20 to fewer for command line plotting
```

### **Seurat PC Heatmaps**

```{r Seurat_heatmaps_1, echo=FALSE,warning=FALSE,message=FALSE}
PCHeatmap(so, pc.use = 1:2, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

```{r Seurat_heatmaps_2, echo=FALSE,warning=FALSE,message=FALSE}
PCHeatmap(so, pc.use = 3:5, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

```{r Seurat_heatmaps_3, echo=FALSE,warning=FALSE,message=FALSE}
PCHeatmap(so, pc.use = 6:8, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

```{r Seurat_heatmaps_4, echo=FALSE,warning=FALSE,message=FALSE}
PCHeatmap(so, pc.use = 9:11, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

```{r Seurat_heatmaps_5, echo=FALSE,warning=FALSE,message=FALSE}
PCHeatmap(so, pc.use = 12:14, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

```{r Seurat_heatmaps_6, echo=FALSE,warning=FALSE,message=FALSE}
PCHeatmap(so, pc.use = 15:17, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

```{r Seurat_heatmaps_7, echo=FALSE,warning=FALSE,message=FALSE}
PCHeatmap(so, pc.use = 18:20, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
```

### **Seurat PCElbowPlot**

```{r Seurat_elbow, echo=FALSE,warning=FALSE,message=FALSE}
PCElbowPlot(so)
```

```{r Seurat_save_rds, echo=FALSE,warning=FALSE,message=FALSE}
gdata::keep(so, params, species, projectId, isHuman, isMouse, sure=TRUE) ## memory optimizations
saveRDS(so, paste0(projectId,"_initial_seurat_object.rds")) ## output SO
exprs <- as.matrix(so@scale.data) ## only need expression matrix for next step, use scaled values
gdata::keep(exprs, params, species, isHuman, isMouse, sure=TRUE)
```

### **GSVA Heatmap - turned off for now**

```{r gsva_heatmap, echo=FALSE,warning=FALSE,message=FALSE,results='hide',fig.keep='all',fig.height=10}
#THIS CAN ONLY BE RUN ON HELIX/BIOWULF (references msigdb files)
#If running locally, please point to different gmt files or remove this chunk

# if(isHuman|isMouse){
#   eset <- ExpressionSet(assayData=exprs)
#   eset_filtered <- varFilter(eset,var.func=IQR, var.cutoff=0.5, filterByQuantile=TRUE)
#   if(isHuman){
#     gsets <- getGmt("/data/CCBR_Pipeliner/db/PipeDB/MSigDB/c2.cp.kegg.v6.0.symbols.gmt")
#   }
#   if(isMouse){
#     gsets <- getGmt("/data/CCBR_Pipeliner/db/PipeDB/MSigDB/c2.cp.kegg.v6.0.symbols_mouse.gmt")
#   }
#   scores <- gsva(exprs(eset_filtered),gsets,min.sz=5,max.sz=500,verbose=TRUE,parallel.sz=1)
#   myscores <- scores[unlist(lapply(rownames(scores),function(x){grepl("^REACTOME_",x)|grepl("^KEGG_",x)|grepl("^BIOCARTA_",x)})),]
#   top <- myscores[names(head(rev(sort((rowSds(myscores)))),50)),]
#   heatmap.2(top,trace="none",labRow=NULL,margins=c(1,20),cexRow=0.7)
# }

#ignore this, just remnants of some stuff I was testing out:
#pc1.scores <- gsva(pc1.exprs,gsets,min.sz=2,verbose=TRUE)
#pc1.cells <- unname(so@dr$pca@cell.embeddings[,1])
#col <- ifelse(pc1.cells<0,"blue","red")
#heatmap.2(pc1.scores,trace="none",labRow=NULL,margins=c(1,20),cexRow=0.7,ColSideColors=col)
```

